% !TEX TS-program = Knitr
\documentclass[a4paper,12pt,times,print,custombib,custommargin]{../Classes/PhDThesisPSnPDF}
%\documentclass[a4paper,12pt,times,numbered,print,index]{PhDThesisPSnPDF}

\usepackage{import}

\input{../Preamble/preamble}

%******************************** Glossary *********************************************
\input{../Glossary/glossary}

%%%%%% -- KNITR SETUP -- %%%%%%%%%%
<<setup_ch2, include=FALSE, cache=FALSE>>=
library(knitr)
# set global chunk options for this child doc
opts_chunk$set(
	fig.align='center', 
	fig.show='hold'
	)
options(replace.assign=TRUE,width=90)
# set up working directory for all code chunks, obviates the need to
# setwd() in each chunk
opts_knit$set(root.dir='~/Data_analysis/SNP-indel-calling/data')
# read in external R code file
read_chunk('../Data_analysis/SNP-indel-calling/data/PopGen-with-geno-uncertainty.R')
# include preamble from main doc for isolated compilation
# doesn't currently work because relative paths are used there
# set_parent('../thesis.Rnw')
@
%************************************************************************************************


\begin{document}

\tableofcontents
%
%\listoffigures
%
%\listof{cmd}{List of commands}
%
%%\todototoc
\listoftodos

\printglossaries


% **************************** Define Graphics Path **************************
%% Note, every path needs to end with a "/" !!!
\ifpdf
    \graphicspath{
    {./Figs/Raster/}
    {./Figs/PDF/}
    {./Figs/}
    {../Data_analysis/SNP-indel-calling/figure/}
    {/Users/Claudius/Documents/PhD/THESIS/kks32/LaTeX/Presentations/PEB_Bialowieza/figures/}
    {/Users/Claudius/Documents/PhD/THESIS/kks32/LaTeX/Data_analysis/reference-mapping/data/simulation/Mapping_Tool_Test_files/figure-html/}
    {/Users/Claudius/Documents/PhD/THESIS/kks32/LaTeX/Data_analysis/SNP-indel-calling/BAM/Big_Data_files/figure-markdown_github/}
    }
\else
    \graphicspath{ 
    {./Figs/Vector/}
    {./Figs/}
    {/Users/Claudius/Documents/PhD/THESIS/kks32/LaTeX/2_Chapter/Figs/Raster/}
    }
\fi

%
%
%
\chapter{Investigation into the demographic history of the hybrid zone}
%
%
%
\begin{figure}[htb]
\centering
\includegraphics[width=.9\linewidth]{grasshopper}
\caption{\textit{Chorthippus parallelus parallelus}. Male individual from Finland.}
\label{Fig:grasshopper}
\end{figure}


%
%\begin{comment}
%

%\abstract{
%In this chapter I want to use all RAD data from all different libraries that I created to investigate the evolutionary history of the hybrid zone.
%
%I will first need to assemble the RAD reads of the different libraries. For that, I want to look into and test a few different programmes, pipelines and analysis protocols that have recently been published. I have created a list of 14 priority papers in my literature library in JabRef under the category \textsf{sRAD} $\rightarrow$ \textsf{RAD de novo assembly}.
%
%I will need to estimate allele frequencies from the RAD data for the two populations I got data for and use it in an ABC analysis. Parameters of interest would be:
%
%\begin{itemize}
%\item time of separation
%\item recent migration rate / rate of gene flow 
%\item recent bottlenecks
%\item signals of selection
%\end{itemize}
%
%FST: \cite{Keinan2007}, \cite{Weir2005}, \cite{McVicker2009}\\
%
%Bias: \cite{Han2014}, \cite{Andrews2016}, \cite{Gautier2012}, \cite{Luca2011}, \cite{Nielsen2011}, \cite{Crawford2012a}, \cite{Johnson2008}, \cite{Albrechtsen2010}\\
%
%2D--SFS: \cite{Gutenkunst2009}
%
%Ascertainment of SNP's in different populations. Par should have undergone recent expansion, which would have led to an increase in low frequency variants that are invariant in ERY. Fst would be close the allele frequency in PAR, i. e. low. That means ascertaining SNP's in PAR should produce a lower genome-wide Fst estimate than when ascertaining SNP's in ERY.
%
%I will need to create a list of papers to read about the technical details of ABC analysis. For the implementation, I will need the assistance of Ludovic.
%
%Finally, since I am trying to extend the scarce knowledge about the history of the hybrid zone, I will need to write a little literature review about what is known so far. Most of these papers should be there as printed copies in the hybrid zone paper box. 
%
%%\newcommand{\Todo}[1]{\todo[size=\normalsize,nolist,inline]{#1}} % don't know why \renewcommand doesn't work
%\newcommand{\Todo}[2][orange]{\todo[size=\normalsize,nolist,inline,color=#1]{#2}} % don't know why \renewcommand doesn't work
%
%\begin{itemize}
%\item RAD de novo assembly:
%	\begin{itemize}
%	\item \Todo[green]{create list of papers to read}
%	\item \Todo[green]{read through list of papers and select programmes to test}
%	\item \Todo{test selected programmes and pipelines}
%	\item \Todo{Get an estimate of the sequencing error rate from the standard RAD data.}
%	\item \Todo{run \texttt{RepeatMasker} on Radome}
%	\item \Todo{screen Radome for \textit{Wolbachia} sequences, maybe by doing a proper PE read assembly again with a proper blast annotation}
%	\item \Todo{assemble different libraries}
%	\item \Todo{estimate allele frequencies of SNP's and indels in a Bayesian framework}
%	\end{itemize}
%\item ABC:
%	\begin{itemize}
%	\item \Todo{explore \texttt{ms}}
%	\item \Todo{create list of papers to read}
%	\item \Todo{test selected programmes and pipelines}
%	\item \Todo{account for biases in the data: null alleles}	
%	\item \Todo{design ABC analysis: parameters}
%	\item \Todo{familiarise yourself with Ludovic's ABC pipeline on Iceberg}
%	\item \Todo{do ABC analysis}
%	\item \Todo{analyse posterior distributions of parameters}
%	\end{itemize}
%\item Literature review of hybrid zone history:
%	\begin{itemize}
%	\item \Todo{create list of papers to read}
%	\item \Todo{read and digest papers while writing the review}
%	\end{itemize}
%\end{itemize}
%}

%
%\end{comment}
%


\section{Introduction}

The two subspecies of the flightless grasshopper \textit{Chorthippus parallelus}, \textit{C. p. parallelus} and \textit{C. p. erythropus}, are thought to have come into secondary contact in the Pyrenees between France and Spain after the last Ice Age some 9000 years ago \citep{Hewitt1990}. They form hybrid zones across the cols of this mountain range \citep{Butlin1985,Butlin1985a}. 
Laboratory generated F$_{1}$ male hybrids between \textit{C. p. parallelus} and \textit{C. p. erythropus} from outside the hybrid zone are almost completely sterile with degenerate testes and a severely disrupted meiosis \citep{Hewitt1987}. Analysis of backcross hybrids in the laboratory suggests that a few loci of large effect contribute to hybrid sterility \citep{Llewellyn2008}. 

Males collected from the center of the hybrid zone, however, are almost completely fertile \citep{Ritchie1992}. The absence of sterile males in the center of the hybrid zone is likely due to the reconstruction of ancestral genotypes caused by selection against combinations of derived alleles in hybrids \citep{Butlin1998a}. The two subspecies are polymorphic for these negative epistatic interactions \citep{Shuker2005} and thus compatible (i.e. ancestral) alleles can reach the center of the hybrid zone. Therefore, selection is likely to restrict the introgression of those parts of the genome that are responsible for genic incompatibilities between the genomes of the two subspecies, whereas neutral loci should introgress to some extent in the two subspecies due to dispersal and recombination. 

%\cite{Butlin1998a} and \cite{Shuker2005} have shown that clines for sterility, cuticular hydrocarbons and female preference are steep in contrast to clines of several morphological characters. This indicates differential introgression across the hybrid zone. We therefore expect to find a marked increase in F$_{ST}$ in comparisons between parental populations, and steeper clines at those loci that are involved in Dobzhansky-Muller incompatibilities observed in male hybrids as compared to neutral loci.

Traditionally, population genetic analyses relied on a genotype calling step. However, with low to medium coverage ($<10\times$) next generation sequencing data there can be substantial uncertainty in genotype inferences. The major sources of error are assembly and mapping errors \citep{Li2011} as well as base call errors and the random sampling of allele copies at hererozygous loci. If this uncertainty is ignored during downstream analyses by using the most likely genotype given the sequencing read data at a locus from an individual -- and usually some hard filtering of putatively variant sites based on quality scores or \gls{LRT} p-values -- this can lead to errors or biases in population genetic inferences \citep{Han2014,Crawford2012a, Johnson2008}. Methods that avoid genotype calling by incorporating genotype uncertainty in downstream analyses can largely avoid these biases \citep{Nielsen2012, Li2011}. In addition, by avoiding SNP calling and instead incorporating uncertainty in polymorphic sites, these methods can also avoid ascertainment bias in downstream analyses \citep{Albrechtsen2010}. 

The following study is an attempt to infer some details of the demographic history of the two subspecies from patterns of genetic variation observed in next--generation sequencing data from a reduced representation library.

%
%
%
\section{Materials \& Methods}
%
%
%
%
%
\subsection{Sampling and library preparation}

Individuals from each of the two distal populations of a transect through the eastern part of the Pyrenees Mountains (Col de la Quillane) were sampled by Jamie Hutchison and Roger Butlin in 2008 and preserved in ethanol until DNA extraction in 2009. A \gls{rad} library was prepared according to the protocol of \cite{Baird2008} (see Appendix~\autoref{ch:sRAD_protocol} on page~\pageref{ch:sRAD_protocol} for details).\todo{refer to outline of different RAD protocols in first chapter; this can only be done when the whole thesis gets compiled} ~It was then sequenced on 4 lanes of an Illumina GAIIx with 51 base pair paired-end sequencing which yielded a total number of 52,872,783 read pairs.
% -- Greixer in Spain and Aunat in France -- 
%
\begin{figure}[htb]
\centering
\includegraphics[width=.9\linewidth]{map-1}
\caption{Map of sampling locations. \textsc{jh34-AU}: \textit{parallelus}; \textsc{jh30-GR}: \textit{erythropus}. \textsc{sterility centre}: marks the centre of the cline of sterility as determined by \cite{Shuker2005}. {\scriptsize Details about the creation of this map are provided in the \nameref{sec:suppl_material_methods}.}
}
\label{Fig:sampling-sites-map}
\end{figure}
\todo{move the description of map creation to supplementary M \& M}
%

%
%
\subsection{Read filtering}
%
%
Raw sequencing reads were split by individual barcodes and filtered for base call quality with \texttt{process\_radtags} from \texttt{Stacks} \citep{Catchen2011}. Since all used barcode sequences differed from each other by at least 3 mismatches, \gls{se} reads were allowed to start with a 5 base pair sequence that differed from one of the barcodes used by 1 mismatch. Read pairs were discarded if they did not contain the remainder of the SbfI restriction site with at most one mismatch following the barcode in the \gls{se} read. Base call quality filtering was done with a sliding window across the reads. A read pair was discarded if the average Phred-scaled base call quality in a window dropped below 20. This read pair filtering discarded 8,505,427 raw read pairs (16\%) leaving 44,367,356 read pairs for the analysis. After removal of the 5 bp barcode sequence, the \gls{se} reads were 46 bp long.
%
%
\subsection{De novo assembly}
%
%
In order to create a reference sequence for the RAD tags in this library, I used an assembly strategy very similar to the one implemented in \href{https://github.com/jpuritz/dDocent}{\texttt{dDocent} }\citep{Puritz2014a} (see fig. \ref{Fig:DeNovoAssembly_sketch}). I used only non-redundant sequence reads for de novo assembly of a "RADome". I used \href{https://github.com/gui11aume/starcode}{\texttt{starcode}} (commit 1034408ca6) \citep{Zorita2015} to collapse \emph{read pairs} from each individual separately into unique representatives allowing for an \gls{edit distance} of up to 2. These unique representatives are the canonical sequences from all \glspl{connected component} of the graph from an \gls{all pairs} search. Canonical sequences are chosen by highest read count, then by highest number of connections with other reads. This reduced the number of quality filtered read pairs for de novo assembly to 9,179,521 (\Sexpr{round(9179521/ 44367356*100, digits=0)}\% of original read pair number). Due to the shotgun-type nature of \gls{pe} reads, read pairs should only be identical (up to 2 mismatches) if they come from the same \gls{fragment} of shearing, i. e. if they are PCR duplicates.

After combining all \gls{se} sequences from these unique sequence pairs from all individuals into one fasta file, I have used \href{https://github.com/torognes/vsearch}{\texttt{Vsearch}} (commit 1116d6167b) \citep{Rognes2016} with its subcommand \texttt{cluster\_fast} for heuristic clustering with a pairwise identity threshold of 0.8 of a query sequence with the centroid of a growing cluster. Identity is defined as 1 minus \gls{edit distance}, but excluding terminal gaps. Thus an indel between two reads will only contribute once towards their edit distance.

\begin{figure}[htb]
\centering
\includegraphics[width=.9\textwidth]{DeNovoAssembly_sketch}
\caption{Overview of the de novo assembly strategy.}
\label{Fig:DeNovoAssembly_sketch}
\end{figure}

Next, I have used the \texttt{uclust} format output of \texttt{Vsearch} together with the collapsed read pairs from \texttt{starcode} to create a clustering output in the format of the output from \href{https://sourceforge.net/projects/bio-rainbow/files/}{\texttt{rainbow cluster} }\citep{Chong2012}. I then ran \texttt{rainbow div} on this cluster file. This programme recursively splits the initial clusters into putative alleles/haplotypes while keeping track of the relationships between split clusters in a tree--like data structure. These split clusters are next merged back recursively along this tree structure with \texttt{rainbow merge} that uses the similarity between \gls{pe} reads to determine whether two split clusters should be merged. Remember that the standard RAD protocol \citep{Baird2008} includes random shearing of restriction fragments. This produces \gls{pe} reads of variable distances from the same RAD tag \citep{Etter2011}. The unique feature of \texttt{rainbow} is to be able to utilise these shotgun--type \gls{pe} reads to distinguish alleles from paralogs. In addition, after recursively merging split clusters into putative RAD loci, \texttt{rainbow merge} also performs a local de novo assembly with the sequences of each merged cluster using \gls{se} and \gls{pe} reads. From the assembly output I then extracted the \gls{se} RAD tag sequences together with their longest \gls{pe} read contig (separating each with a sequence of 10 N's) using \texttt{gawk} code from \texttt{dDocent}. 

Finally, since the merge process of \texttt{rainbow} depends on overlap between \gls{pe} reads from the same RAD locus, it could be incomplete when coverage is low. I therefore clustered the \texttt{rainbow} contigs again with the \texttt{Vsearch} subprogramme \texttt{cluster\_fast}, this time with an identity threshold of 0.9 and producing a majority consensus sequence from a multiple alignment of the sequences in each cluster of \texttt{rainbow} contigs. The assembled RADome contains 583,312 contigs and a total length of 97 \gls{Mbp}. In the following I will frequently refer to it under the informal name of \emph{Big Data} reference assembly.

An extensive documentation about the assembly procedure as well further downstream analyses can be found in the \texttt{BASH} script \href{https://github.com/claudiuskerth/PhDthesis/blob/master/Data_analysis/SNP-indel-calling/assembly.sh}{\texttt{assembly.sh}}. This text file contains exact command lines used together with extensive explanation. It should be the first stop for information when trying to reproduce the results presented here. Note, not all steps of the analysis could be documented directly in this text file, but when \href{http://rmarkdown.rstudio.com}{Rmarkdown} or \href{https://ipython.org}{IPython} notebooks \citep{Perez2007} have been used for analysis then references to those file can be found in \texttt{assembly.sh}.

 
%
%
\subsection{Read mapping}
%
%
I mapped all quality filtered reads against the Big Data reference with the programme~\href{http://bowtie-bio.sourceforge.net/bowtie2/manual.shtml}{\texttt{bowtie2}} version 2.2.9 (downloaded on 29th October 2016) \citep{Langmead2012}. I ran \texttt{bowtie2} in end--to--end mode, i. e. without soft--clipping of query sequences. Further, I specified "very--sensitive" search mode with a seed length of 20. I allowed gaps to be up to 10 bp long, but I disallowed gaps within 4 bp of either end of the read. In addition I specified a high penalty of 10 for alignments against ambiguous characters in the reference. In each reference contig, \gls{se} RAD tag and \gls{pe} contig were separated by 10 N's. This high penalty therefore makes successful read alignments across the gap between \gls{se} RAD tag and \gls{pe} read contig very unlikely. Further, I specified a minimum fragment length of 60 for \gls{proper pair} read alignments, thus allowing some overlap between reads in a pair. 

Although standard RAD data from paired--end sequencing allows the detection and removal of PCR duplicates, I have opted not to do so for convenience. The following analyses all depend on the correct estimation of genotype likelihoods which assumes that reads are sampled independently from the true genotype.\todo[color=green]{I have removed PCR duplicates with starcode and tried to estimate 1D SFS but without success. See \texttt{deduplicated\_spectra.ipynb}.}~Due to the PCR duplicates I expect the genotype likelihoods to be generally biased against heterozygote genotypes.

%
%
\subsection{Filtering}
%
%

There are 4,575 contigs with \gls{se} reads that map to the \gls{pe} part of the contig, i. e. which do not pass the \textit{mismapping} filter. I have also filtered \glspl{contig} for excessive coverage by \gls{se} reads, which map to the \gls{RAD tag} part of each contig (i. e. next to the restriction site). The coverage by \gls{pe} reads is much more dependent on the length of the contig. With this filter a contig has excessive coverage if in any of the individual BAM files it has \gls{se} read coverage above the 99th percentile of that individual's coverage distribution. I have thus excluded 2,282 contigs due to excessive coverage. Since many false--positive \glspl{snp} fall within low--complexity sequences \citep{Li2014}, I have detected those regions within the \emph{Big Data} reference assembly with \texttt{dustmasker} \citep{Morgulis2006} and excluded them from further analysis. The above filters have excluded 10\% of the reference from further analysis, leaving 88 \gls{Mbp}. %before filtering: sum of interval length from Big_Data_ref.fa.bed, after filtering: sum of interval length from Big_Data_Contigs.noSEgt2.nogtQ99Cov.noDUST.bed

I then went on to extract those bits of the filtered reference that have sufficient data for downstream analyses. I used the \texttt{samtools depth} command to get the read count for all positions and for all 36 individuals across the filtered reference. I then extracted those positions where at least 15 individuals have each at least 3$\times$ coverage. I only counted reads with a mapping quality greater than 5. This reduces the number of sites to 2\% of the original \emph{Big Data} reference assembly or 2 \gls{Mbp}, spread over 34,967 contigs. The first 6 base pairs of each contig are the remainder of the SbfI restriction site (\texttt{TGCAGG}) and therefore not variable. I have excluded those sites as well from further analysis. 

In addition to the excessive coverage filter, I have also applied a HWE filter to the remaining sites in order to reduce false heterozygotes caused by reads from paralogous sequences mapping to the same position in the reference. I have used the programme \texttt{ANGSD}\todo{which version, which git commit} \citep{Korneliussen2014} to estimate per site inbreeding coefficients ($F_{site}$) as well as LRT p--values using genotype likelihoods \citep{Vieira2013}. I removed entire contigs that had a site with a negative estimate of inbreeding coefficient and a p--value for deviation of $F_{site}$ from 0 of less than 0.05.  I estimated inbreeding coefficients for each population separately as well as for both populations taken together. The latter increased sample size for the estimation of $F_{site}$ but also increased estimates of $F_{site}$ for \glspl{snp} with different allele frequencies in the two populations. This filter removed 217 contigs. This has left 1,799,962 filtered sites on 34,750 contigs for further analysis. 
%
\begin{figure}[htb]
\centering
<< global-coverage-dist, out.width='.9\\linewidth', echo=FALSE, cache=TRUE>>=
@
\caption{Histogram of across sample coverage with reads of mapping quality greater than 5 from 1.7 million nucleotide sites in the \emph{Big Data} reference after filtering.}
\label{Fig:global-coverage-dist}
\end{figure}
%
\\
Inspection of the global, i. e. across sample, coverage distribution revealed that there were 12,693 sites with total coverage greater than 1000$\times$. I have therefore also determined the 99th percentile of the global per--site coverage distribution and removed all contigs with a position that had coverage greater than the global 99th percentile. This has removed 69,438 sites on 407 contigs leaving 1,730,524 sites on 34,343 contigs for further analysis. The new coverage distribution is shown in figure \ref{Fig:global-coverage-dist}.
%
\begin{figure}[htb]
\centering
<<mapQ-dist, out.width='.49\\linewidth', echo=FALSE,cache=TRUE>>=
@
\caption{Distribution of mapping quality scores determined by \texttt{bowtie2} for all mapped reads to the unfiltered and filtered reference sequence.}
\label{Fig:mapQ-dist}
\end{figure}
%
\\
The above filters have selected for contigs with more unique sequences as shown by the great reduction in the proportion of reads with a mapping quality score of 1 that map to the filtered reference as compared to the unfiltered reference sequence (see fig. \ref{Fig:mapQ-dist}). Only reads with a mapping quality score of $\ge$ 5 have been used for downstream analyses. A Phred score of 5 for the mapping quality should indicate a probability of about 1/3 that the read truly originated elsewhere in the \textsl{reference sequence} (mapping uncertainty does not incorporate assembly uncertainty). However, as figure \vref{Fig:obs-vs-reported-FPR} shows, \texttt{bowtie2} generally greatly underestimates the true mapping quality, given the reference sequence, particularly for very low mapping quality scores. A more stringent filtering of read mappings is therefore not necessary. Note that in the following analyses mapping quality information is incorporated into the calculation of genotype likelihoods by capping base quality by the mapping quality of the read\footnote{This claim has not been fully verified but is very likely: I have used the genotype likelihood version from \texttt{samtools} throughout (and \texttt{ANGSD} directly incorporates code from \texttt{samtools}), which should be based on \texttt{MAQ} \cite[section Methods: Consensus genotype calling]{Li2008}.}. A different mapping programme will therefore result in different genotype likelihoods as mapping quality scores are computed with different heuristics by the different mapping programmes (see fig. \vref{Fig:obs-vs-reported-FPR-a}). 
%
\begin{figure}[htb]
\centering
<<mappability-dist, out.width='\\linewidth', echo=FALSE, cache=TRUE>>=
@
\caption{Distribution of average mappability per contig before and after filtering the Big Data reference assembly. For details see section \vref{Sec:mappability} in the supplementary material.}
\label{Fig:mappability-dist}
\end{figure}
%
\\
A better measure of the repetitiveness of the assembled RADome than the distribution of read mapping quality is the distribution of mappability. Mappability measures the uniqueness of a sequence of length $k$ (kmer's) sampled from a position $x$ in the reference sequence (here Big Data reference assembly). Mappability is the inverse of how often that kmer can be found in the whole reference sequence up to a specified number of mismatches (or edits) \citep{Derrien2012}. Mappability has a range between 0 and 1. A position has mappability of 1 if its kmer occurs only once in the reference, 0.5 if it occurs twice and so on. Figure \ref{Fig:mappability-dist} shows the distribution of per contig average mappability of the raw Big Data reference assembly and of the reference assembly after applying above filters. It shows that a large proportion of contigs have an average mappability of 1 or close to 1 and that the majority of contigs have average mappability greater than 0.5. It also shows that the above filters have \emph{not} selected for more unique contigs in the Big Data reference assembly. In fact, the reference contigs kept after filtering for further analysis have \emph{lower} median mappability than before filtering. The median mappability before filtering is \Sexpr{round(median(before), 3)} whereas after filtering it is \Sexpr{round(median(after), 3)}. The median of the difference in mappability between a contig before filtering and a contig after filtering is \Sexpr{round(u$estimate, 3)} (95\% \gls{CI}: \Sexpr{round(u$conf.int, 3)}). ~Regions in the reference with lower mappability are more difficult to map reads to with data from short read next-generation sequencing. Quantitative measures of read counts as well as polymorphism detection in those regions will be compromised \citep{Derrien2012,Lee2012a}. However, a lower mappability is not indicative of a bad de novo assembly of the reference. In fact, collapsing repetitive sequences from the sampled genome into one reference contig will \emph{increase} the average mappability of the assembled reference.
\\
%
\begin{figure}[htb]
\centering
<<contig-length-dist, out.width='\\linewidth', echo=FALSE, cache=TRUE>>=
@
\caption{Distribution of contig lengths. Top: raw RAD assembly (\emph{Big Data} reference) before filtering (583,312 contigs). Bottom: RAD assembly after applying filters (34,343 contigs).}
\label{Fig:contig-length-dist}
\end{figure}
%
\\
The above filters have also selected for contigs with lengths that are roughly consistent with the expected size distribution of sonicated restriction fragments that were size selected on an agarose gel during library preparation (see figures \ref{Fig:contig-length-dist} and \vref{shearedPools}). Note that the size selected DNA fragments contained the 67 bp long P1 adapter and that PCR as well as cluster generation on the flow cell should increase the proportion of short fragments in the resulting Illumina sequence data.

The average per site, per individual coverage across the 1.7M filtered sites is \Sexpr{round(cov, digits=1)}$\times$ (fig. \ref{Fig:individual-coverage-dist}).
%
\begin{figure}[htb]
\centering
<<ind-cov-dist, out.width='.9\\linewidth', echo=FALSE,cache=TRUE>>=
@
\caption{Individual coverage distributions from reads with mapping quality greater than 5 over 1.7 million sites in the \emph{Big Data} reference after filtering. red: \textit{erythropus}, green: \textit{parallelus}.}
\label{Fig:individual-coverage-dist}
\end{figure}
%


%
%
\subsection{PCA}
%
%
I have used \texttt{ANGSD}  \citep{Korneliussen2014} to estimate posterior genotype probabilities across all sites using as prior a \gls{ML} estimate of the \gls{population minor allele frequency} (MAF) per site across all individuals from the two populations and the assumption of \gls{HWE} \citep{Kim2011}. Although the assumption of \gls{HWE} may well be violated, it should only make genotype probability estimates more similar between divergent individuals. Therefore, any major divergence detected between individuals based on these genotype probabilities cannot be an artifact of this prior. At low coverage, the identification of the minor allele can be uncertain. The estimation of the minor allele frequency incorporated this uncertainty \cite[suppl. mat.]{Skotte2012}. I have then estimated the \gls{ML} global unfolded \gls{SFS} based on per--site \gls{sample allele frequency} likelihoods (SAF), again across all individuals, using \texttt{ANGSD}'s subprogramme \texttt{realSFS} \citep{Nielsen2012, Korneliussen2013}. This global \gls{SFS} then served as prior to estimate per--site \gls{sample allele frequency} posterior probabilities. Thus for each site $2N+1$ probabilities were calculated (N=36, the total number of all individuals sequenced). Finally, I estimated a covariance matrix with \texttt{ngsCovar} \citep{Fumagalli2014} between all 36 individuals by summing over genotype probabilities and weighting each site by its probability of being variable \citep[eq. 19 and 20]{Fumagalli2013}.\todo{in the same way it should be possible to calculate the covariance matrix between all SNP's, thereby getting composite LD estimates}~ I then performed an eigen--decomposition on the resulting covariance matrix with the \texttt{R} function \texttt{prcomp}.
%
%
\subsection{$F_{ST}$}
%
%
Using \texttt{ANGSD} and its subprogramme \texttt{realSFS} \citep{Korneliussen2014}, I have determined a ML estimate of the global \emph{unfolded} 2D--\gls{SFS} that contains the joint sample allele frequencies of the non--reference allele from \textit{erythropus} and \textit{parallelus}. I required a nucleotide site to have sequence reads from at least 9 individuals in each population. The 2D--SFS contains the joint frequencies from approximately 1.13 million sites. This global unfolded 2D--\gls{SFS} was then used as (empirical) prior for posterior sample allele frequency probabilities from which the posterior expectation of $F_{ST}$ per site is calculated using \texttt{realSFS} from ANGSD \citep{Fumagalli2013}. This programme reports the numerator and denominator for either Reynolds' \citep[eq. 3]{Fumagalli2013} or Hudson/Bhatia's $F_{ST}$ \citep[eq. 9 and 10]{Bhatia2013} for each site. I am not reporting estimates of $F_{ST}$ per site due to their large variance \citep{Weir2005}. Instead I am reporting \emph{average} $F_{ST}$ across sites within a contig or the global average $F_{ST}$ across all sites. This average is always calculated by summing nominator and denominator across sites and then taking the ratio \citep{Bhatia2013}.
In order to investigate the dependence of $F_{ST}$ on SNP ascertainment in one or the other population, I have calculated ML estimates of \gls{population minor allele frequency} (MAF) with \texttt{ANGSD} \citep{Kim2011}. I have then used these allele frequencies to subset sites into minor allele frequency classes and estimated average $F_{ST}$ for each class as before.\todo{these analyses are also documented in assembly.sh. It may be good to create a separate script for analyses with ANGSD/realSFS and then provide a link to it here.}
%
%
\subsection{Genetic diversity estimates}
%
%
I have calculated ML estimates of the \emph{unfolded} global \gls{SFS}, i. e. including all sites across contigs, for \textit{erythropus} and \textit{parallelus} separately with \texttt{ANGSD} and its subprogramme \texttt{realSFS}\todo{which version? note that I folded myself. link to \texttt{assembly.sh}} ~\citep{Korneliussen2014}. 
I required 9 individuals to have read data in order to include a site into the estimation. Since I noticed a considerable amount of variability in the ML estimates of SFS from repeated runs of the programme \texttt{realSFS}, I turned off its accelerated \gls{EM} algorithm, extended the maximum number of \gls{EM} iterations to 50,000 and specified a \textit{tolerance} of 1e-06, which is the minimum difference in likelihood between two successive \gls{EM} iterations and therefore a stopping criterium. These modifications completely removed the variability in SFS estimates, but also greatly extended running time, particularly for the \gls{SFS} of \textit{parallelus}. The programme \texttt{realSFS} allows bootstrapping the \gls{SFS}, i. e. re--estimation of the \gls{SFS} from nucleotide sites resampled with replacement. However, \gls{CI} estimated from these bootstraps will be too narrow, since it ignores that sites in the same contig are not independent and the SFS estimated is a \emph{composite} maximum likelihood estimate. I have therefore created 200 bootstrap replicates of contigs and estimated site frequency spectra for these as before.\todo{document bootstrapping over contigs.}

I fitted standard-neutral-model site frequency spectra to the observed folded single-population spectra by optimising $\theta$ in the following equation \citep[eq. 4.21]{Wakeley2009}:
 \begin{equation}
E[\eta_i] = \theta \frac{\frac{1}{i} + \frac{1}{n-i}}{1+\delta_{i,n-i}} \qquad 1 \le i \le \big[n/2\big]
\end{equation}
This formula gives the neutral expectation of counts ($\eta$) in each frequency class (i) in a folded spectrum. I used the R\todo{cite R, add link to Rmd file}~function \texttt{optimize} to find the value of $\theta$ that minimizes the deviation of the above equation from all observed counts $\eta_i$. 
%According to \cite{Fu1995}, p. 192, for large sample sizes (i. e. large $n$) the counts in neutral spectra can be approximated by a Poisson distribution. I therefore added 95\% \gls{CI} limits from Poisson distributions with parameters equal to $E[\eta_i]$ to the fitted neutral spectra.

I derived the global estimates of the number of segregating sites ($S$) and the average number of pairwise differences ($\pi_{Tajima}$) from the global folded \gls{SFS} according to equations 1.3 and 1.4 in \cite{Wakeley2009}, respectively:
%
\begin{equation}
S = \sum_{i=1}^{n/2} \eta_{i}
\end{equation}
%
\begin{equation}
\pi = \frac{1}{\binom{n}{2}} \sum_{i=1}^{n/2} i(n-i)\eta_i
\end{equation}
%
where $n$ is the number of alleles sampled, i. e. twice the number of sampled individuals for diploid loci. $\eta_i$ is the count in the i--th frequency class of the folded SFS.
The estimates of global Tajima's D are based on these estimates and equation 2.17 in \cite{Gillespie2004}:
%
\begin{equation}
D_{T} = \frac{\pi - \theta_{W}}{C}
\end{equation}
$\theta_{W}$ is Watterson's $\theta$ and is calculated with:
\begin{equation}
\theta_{W} = \sfrac{S}{\sum_{i=1}^{n-1} \frac{1}{i}}
\end{equation}
Details of the above calculations are documented in the Rmarkdown notebook \href{https://bitbucket.org/claudiuskerth/phdthesis/src/fc097ea0cfc02ee2ed3d2d53a6c6ec40deef433a/Data_analysis/SNP-indel-calling/ANGSD/SFS/SFS.Rmd?at=master}{\texttt{SFS.Rmd}}.

The estimates for \textit{parallelus} are based on 1,214,939 sites, the estimates for \textit{erythropus} are based on 1,638,468 sites.\todo{I also determined these statistics from only overlapping sites.} 

%I have also estimated Tajima's D per contig with the empirical Bayes method described in \cite{Korneliussen2014}. This uses the ML estimate of the global folded \gls{SFS} as prior for posterior sample allele frequency probabilities that are then summed across the sites in each contig to get the per--contig estimate of \gls{SFS}.

%
%
%
\section{Results \& Discussion}
%
%
%

%
%
\subsection{Genetic differention between \textit{erythropus} and \textit{parallelus}}
%
%
Figure \ref{Fig:PCA} shows a principal component analysis of genotypic covariances between all 36 individuals of \textit{erythropus} and \textit{parallelus} taking SNP and genotype uncertainty into account. Almost $^1/_4$ of the total genotypic variance is explained by the first principal component which clearly separates the two subspecies. The second component captures only 3\% of the total genotypic variance. This confirms estimates of genetic differentiation by \cite{Cooper1995} -- based on sequence variation at a single nuclear locus -- that the two subspecies are genetically distinct. Individual \texttt{par\_34-1} seems to be considerably less differentiated from \textit{erythropus} than the other \textit{parallelus} individuals. This may be due to insufficient information because of extremely low coverage (see fig. \ref{Fig:individual-coverage-dist})
%
\begin{figure}[htb!]
\centering
<< PCA, out.width='.9\\linewidth', echo=FALSE,cache=TRUE>>=
@
\caption{Principal component analysis with genotype probabilities across 1.7 million sites from all 36 individuals of \textit{erythropus} and \textit{parallelus}. Each site was weighted by its probability of being variable during the estimation of genotypic covariances between individuals.}
\label{Fig:PCA}
\end{figure}
%

I estimated posterior expectations of $F_{ST}$ from 1.6 million sites across 32,706 contigs. Note, that this included 0.5 million sites for which there were fewer than 9 individuals with read data in one of the two populations.
%
\begin{figure}[htb]
\centering
<< permut-global-fst-hist, out.width='.9\\linewidth', echo=FALSE, cache=TRUE>>=
@
\caption{Resampling based estimation of the variance of global $F_{ST}$ (blue) and distribution of global $F_{ST}$ when randomly permuting individuals into two populations (green).}
\label{Fig:global-fst}
\end{figure}
%
The global average Hudson/Bhatia's estimate of $F_{ST}$ is \Sexpr{signif(Fst.bhatia.global, digits=3)} with a bootstrap 95\% \gls{CI} between \Sexpr{signif(bhatia.CI95[1], 3)} and \Sexpr{signif(bhatia.CI95[2], 3)} (see fig. \ref{Fig:global-fst}).\todo{mention how you did the bootstrapping, documented in Fst.Rmd} ~Note, that the 95\% \gls{CI} only captures the variance due to sampling loci in the genome (\emph{genetic sampling} according to \cite{Weir1996}, p. 161), not the variance due to sampling individuals from the population. The latter could be estimated by bootstrapping individuals from each population. \todo{read \cite{Holsinger2009}}~Estimating \gls{CI}'s by bootstrapping over contigs assumes that they are independent replicates of the evolutionary process in the history of the two subspecies, i. e. that they are unlinked. However, with the standard RAD protocol \citep{Baird2008}, \emph{two} RAD tags (here called contigs) are recovered from each restriction site. The bootstrap \gls{CI} for genome-wide $F_{ST}$ computed here may therefore slightly underestimate uncertainty. Also note that the null distribution of global $F_{ST}$, estimated by randomly permuting individuals into two populations, does not include 0. There therefore seems to be a positive bias in the estimation of global $F_{ST}$ of about \Sexpr{signif(perm.fst.CI95[2], 3)}.\todo{what about Reynold's Fst? Does it show the same bias? estimate Fst from 2D SFS.}

Figure \ref{Fig:fst-by-ascert-class} shows the dependence of average $F_{ST}$ on the minor allele frequency when only sites are included that are polymorphic in the ascertainment population. Not surprisingly, SNP ascertainment in one of the two populations lowers average $F_{ST}$ estimates since it also excludes highly differentiated sites where the ascertainment population is fixed for one allele.
%
\begin{figure}[htb]
\centering
<< fst-by-ascertainment-class, out.width='.9\\linewidth', echo=FALSE, cache=TRUE>>= %out.height='.4\\textheight', 
@
\caption{$F_{ST}$ by ascertainment class. The global average $F_{ST}$ when using all sites, i. e. without SNP ascertainment, is shown as a grey line.}
\label{Fig:fst-by-ascert-class}
\end{figure}
%
When ascertaining SNP sites in \textit{erythropus} or \textit{parallelus}, in each case the average $F_{ST}$ estimate tends to decrease with decreasing minor allele frequency. This is consistent with both populations having undergone a recent population expansion where rare alleles are more likely to be private to one population. This is made evident by the estimate of global unfolded 2D-\gls{SFS} that was used as prior for per--site $F_{ST}$ estimates (fig. \ref{Fig:unfolded-2D-SFS}). Since these sites have a low allele frequency differentiation, they lower the average $F_{ST}$ estimate. In contrast, if one or both populations had undergone a recent population bottleneck, variable sites would more often be ancient and polymorphic in the ancestral population. \todo{Richard has suggested an alternative explanation, which is due to the higher variance of $F_{ST}$ for loci with low total heterozygosity (which is equivalent to low minor allele frequency in the total sample). He suggested using simulated 2D spectra under models of population expansion and constant population size to verify my claims. \cite{Albrechtsen2010}}
Sites with rare alleles in the bottlenecked population would therefore also include those that have drifted apart in allele frequency between the two populations, thus increasing average $F_{ST}$ estimates \citep{Bhatia2013}.\todo{see also \cite{Roesti2012a}}
%
\begin{figure}[H]
\centering
<< unfolded-2D-SFS, out.width='.9\\linewidth', echo=FALSE, cache=TRUE, warnings=FALSE>>=
@
\caption{ML estimate of the joint sample allele frequencies of the non--reference allele for \textit{parallelus} (PAR) and \textit{erythropus} (ERY). The joint frequency class (0,0) has been set to zero for the purpose of better visualisation.}
\label{Fig:unfolded-2D-SFS}
\end{figure}
%

A ML estimate of the number of nucleotide sites with fixed differences between these two samples of 18 individuals each of \textit{erythropus} and \textit{parallelus} can also be extracted from the unfolded 2D--\gls{SFS}: \Sexpr{signif(sfs2d[1,37] + sfs2d[37,1], 3)} or \Sexpr{signif((sfs2d[1,37] + sfs2d[37,1])/sum(sfs2d)*100, 2)}\% of polymorphic sites. 
{\itshape
\cite{Whitlock2015} have recently described a method to detect loci likely under spatially heterogeneous selection by modelling the neutral distribution of $F_{ST}$ for specific demographic scenarios. Figure \ref{Fig:fst-by-contig-hist} shows the observed distribution of $F_{ST}$ across all contigs. The method of \cite{Whitlock2015} could be applied to this distribution for inference of likely demographic scenarios.
}\todo{This could be moved to the section Future directions on page \pageref{sec:future_directions}.}
%
\begin{figure}[htb]
\centering
<< fst-by-contig-hist, out.width='.9\\linewidth', echo=FALSE, cache=TRUE>>=
@
\caption{Distribution of $F_{ST}$ across contigs.}
\label{Fig:fst-by-contig-hist}
\end{figure}
%
\clearpage
%
%
%
\subsection{Genetic diversity within \textit{erythropus} and \textit{parallelus}}
%
%
%
Figure \ref{Fig:folded-sfs} shows the \gls{ML} estimate of the global folded \gls{SFS} for \textit{erythropus} and \textit{parallelus}. The conspicuous spike in frequency for the minor allele count of 2 in \textit{parallelus} is not compatible with any demographic scenario.\todo[color=green]{add expected neutral SFS with optimized $\theta$. See eq. 4.21 in \cite{Wakeley2009} and \texttt{dadi\_exercises.ipynb} as well as \cite{Fu1995, Fu1994}.} ~However, even when disregarding frequency class 2, the folded \gls{SFS} of \textit{parallelus} seems to be more right--skewed, that is having a greater proportion of low-frequency variants, than \textit{erythropus}. 
%
\begin{figure}[htb]
\centering
<< folded-sfs, out.width='\\textwidth', echo=FALSE, cache=TRUE>>=
@
\caption{ML estimate of the global minor allele frequency spectrum for \textit{erythropus} (red) and \textit{parallelus} (green). The site frequency spectra were estimated from either \textsl{only overlapping} nucleotide sites (1.13 million) or \textsl{including non-overlapping} sites (1.6 million from \textit{erythropus} and 1.2 million from \textit{parallelus}).}
\label{Fig:folded-sfs}
\end{figure}
%
\todo{what means \textsl{overlapping sites}?}
Note, that this \gls{SFS} is not based on called \glspl{snp}. Instead it is based on per-site sample allele frequency likelihoods (SAF) that incorporate the genotype uncertainty of each individual due to binomial sampling of alleles from a diploid genotype -- as happens during next generation sequencing -- as well as sequencing and read alignment error\todo[color=green]{really? Also read alignment error? Yes! see cite{Li2008}} ~\citep{Li2008, Li2011, Nielsen2012}.
\todo[inline]{Could the inflated $\eta_2$ in both populations be due to singletons on the haploid X chromosome of these male grasshoppers? How big is the X chromosome with respect to the rest of the genome? $\rightarrow$ fourth largest of 9 chromosomes (see Gosalvez1988, fig. 2);  How does ANGSD deal with haploid loci? $\rightarrow$ the genotype likelihoods refer to diploid loci; Does ANGSD's ML estimation of SFS include HWE expectation? $\rightarrow$ yes, SAF likelihood calculation includes the assumption of HWE;  Is there a way to specifically detect contigs from the X chromosom? $\rightarrow$ see \texttt{1D\_model\_synthesis.ipynb}, section 2.2 "reducing noise and bias"} 

~It does not therefore suffer from the ascertainment bias observed in previous studies that were based on sites detected as polymorphic in a specific sample \citep{Han2014,Albrechtsen2010, Korneliussen2013}. This ascertainment bias has led to a relative excess of intermediate versus low frequency classes\todo{and genotype calling generally leads to an excess of low frequency variants, see fig. 1 in \cite{Nielsen2012}}. 

The \gls{SFS} has however not yet been corrected for bias due to allele--drop--out \citep{Luca2011, Cariou2016}. 
\begin{quotation}
simulations indicate this bias is of minor import- ance when the polymorphism is below 2 \%, which is the case in most species, at least in animals
\end{quotation}
\citep{Cariou2016} \comment{Find estimates of $\pi$ for African Drosophila in \cite{Pool2012} and a comparison of diversity estimates across several species in \cite{Romiguier2014} for reference.}
Allele--drop--out is a problem that has affected previous types of genetic markers like microsatellites or AFLP's, usually known under the term "null alleles".~\todo{check for the effect of allele-drop-out on the SFS. see also \cite{Davey2012}}
\cite{Arnold2013} and \cite{Gautier2012} have shown with simulations that loci affected by allele--drop--out show a greater proportion of intermediate allele frequencies as compared to loci not affected. They therefore show a relative excess of observed nucleotide heterozygosity ($\pi_{Tajima}$). My data filtering, that included filtering for minimum coverage and for minimum number of individuals with read data, should have enriched for sites less affected by allele--drop--out. Those sites show the opposite bias towards a relative excess of low--frequency variants \citep{Arnold2013}. Since the same filtering thresholds were applied to the data from \textit{erythropus} and \textit{parallelus} and since the \textit{parallelus} sample has generally much lower coverage (see fig. \ref{Fig:individual-coverage-dist}), it seems plausible that the sites used to estimate the \gls{SFS} for \textit{parallelus} have been relatively more enriched for those less affected by allele--drop--out. This may explain the relative excess of low--frequency variants in \textit{parallelus} compared to \textit{erythropus}.\comment{I have estimated $\pi$ from marginal 1D spectra, i. e. from only overlapping sites. The estimates of $\pi_{site}$ change only marginally. I think that rules out that a different degree of allele-drop-out could explain the differences in diversity. See section "marginal spectra" in \texttt{05\_2D\_models.ipynb}.}

The allele--drop--out problem is a missing data problem and to some extent similar to the missing data problem caused by low coverage next--generation sequencing. Both increase false homozygote calls for true heterozygotes and both therefore lead to a deviation of observed genotype frequencies from \gls{HWE}. The algorithms for inferring the \gls{SFS} implemented in \texttt{ANGSD} incorporate an assumption of HWE \cite[eq. 2 and 3]{Nielsen2012}. They place a low conditional probability on combinations of genotypes in the sample that would deviate strongly from HWE. I therefore expect \texttt{ANGSD} to mitigate the effects of allele--drop--out on population genetic analyses.

%\begin{enumerate}
%\item incorporating genotype uncertainty into downstream population genetic analyses
%\item "imputing" genotype likelihoods by estimating the population minor allele frequency using the raw genotype likelihoods from all individuals in the same population and \emph{assuming \gls{HWE}}.\todo{is this really true?}
%\end{enumerate}



%
<< fit-neutral-theta, echo=FALSE, warning=FALSE>>=
@
%

%
\begin{figure}[htbp]
\centering
<< folded-sfs-boot, out.width='.49\\linewidth', echo=FALSE, cache=TRUE, warning=FALSE>>=
@
\caption{The global folded \gls{SFS} of \textit{erythropus} and \textit{parallelus} with absolute counts and 95\% \gls{CI} limits from 200 bootstrap resamples of contigs. A standard neutral model spectrum was fit to each observed spectrum for comparison.}
\label{Fig:folded-sfs-boot-exh}
\end{figure}
%

Figure \ref{Fig:folded-sfs-boot-exh} shows that despite the large amount of sites included in the ML estimation, there is still considerable uncertainty in the global folded \gls{SFS}, particularly for \textit{parallelus}. This could be explained by the lower average sequence coverage for individuals from the \textit{parallelus} population (see fig. \ref{Fig:individual-coverage-dist}).


<< S-and-Pi, echo=FALSE, cache=TRUE, warning=FALSE>>=
@
<< Tajimas-D, echo=FALSE, cache=TRUE, warning=FALSE>>= %out.height='.4\\textheight', 
@

%\Sexpr{formatC(quantile(S_prop.ery.boot, probs=c(.025, .975)), format="f", digits=4)}

%
\begin{table}
\centering\scriptsize
\setlength{\tabcolsep}{7pt}
\caption{Comparison of the proportion of segregating sites $S_{prop}$, the average number of pairwise differences per site $\pi_{site}$ and Tajima's D for the \gls{SFS} of \textit{erythropus} and \textit{parallelus}. Numbers in brackets are 95\% bootstrap confidence intervals.}
\label{Tab:diversity_parameters}
\vspace{5pt}
\begin{tabular}{lllll}
%\begin{tabularx}{\textwidth}{XXXXX}
\toprule
\multirow{2}{*}{parameter} & \multicolumn{2}{c}{\textbf{only overlapping sites}} & \multicolumn{2}{c}{\textbf{including non-overlapping sites}} \\[5pt]
                                    & \textit{erythropus} & \textit{parallelus} & \textit{erythropus} & \textit{parallelus} \\
\midrule
$S_{prop}$ & \Sexpr{formatC(S_prop.ery, format="f", digits=4)} (\Sexpr{formatC(quantile(S_prop.ery.boot, probs=c(.025, .975)), format="f", digits=4)}) & \Sexpr{formatC(S_prop.par, format="f", digits=4)} (\Sexpr{formatC(quantile(S_prop.par.boot, probs=c(.025, .975)), format="f", digits=4)}) & \Sexpr{formatC(S_prop.ery.nonO, format="f", digits=4)} (\Sexpr{formatC(quantile(S_prop.ery.boot.nonO, probs=c(.025, .975)), format="f", digits=4)}) & \Sexpr{formatC(S_prop.par.nonO, format="f", digits=4)} (\Sexpr{formatC(quantile(S_prop.par.boot.nonO, probs=c(.025, .975)), format="f", digits=4)})\\[3pt]
%\rowcolor[gray]{0.9} 
$\pi_{site}$ & \Sexpr{formatC(pi_sites.ery, format="f", digits=5)}  (\Sexpr{formatC(quantile(pi_sites.ery.boot, probs=c(.025, .975)), format="f", digits=5)}) & \Sexpr{formatC(pi_sites.par, format="f", digits=5)} (\Sexpr{formatC(quantile(pi_sites.par.boot, probs=c(.025, .975)), format="f", digits=5)}) & \Sexpr{formatC(pi_sites.ery.nonO, format="f", digits=5)} (\Sexpr{formatC(quantile(pi_sites.ery.boot.nonO, probs=c(.025, .975)), format="f", digits=5)}) & \Sexpr{formatC(pi_sites.par.nonO, format="f", digits=5)} (\Sexpr{formatC(quantile(pi_sites.par.boot.nonO, probs=c(.025, .975)), format="f", digits=5)})\\[3pt]

Tajima's D &  \Sexpr{formatC(ery.TajimasD.global, format="f", digits=3)} (\Sexpr{formatC(quantile(ery.TajimasD.global.boot, probs=c(.025, .975)), format="f", digits=3)}) & \Sexpr{formatC(par.TajimasD.global, format="f", digits=3)} (\Sexpr{formatC(quantile(par.TajimasD.global.boot, probs=c(.025, .975)), format="f", digits=3)})  & \Sexpr{formatC(ery.TajimasD.global.nonO, format="f", digits=3)} (\Sexpr{formatC(quantile(ery.TajimasD.global.boot.nonO, probs=c(.025, .975)), format="f", digits=3)}) & \Sexpr{formatC(par.TajimasD.global.nonO, format="f", digits=3)} (\Sexpr{formatC(quantile(par.TajimasD.global.boot.nonO, probs=c(.025, .975)), format="f", digits=3)})\\
\bottomrule
\end{tabular}
\end{table}
\todo{regarding potential reasons for negative Tajima's D see \cite{Reed2005} and \cite{Corbett-Detig2015a}}
%
%%
%%
%%


The number of segregating sites ($S$) and the average number of pairwise differences ($\pi_{Tajima}$) are summary statistics of the \gls{SFS}. While $S$ weights each site equally, intermediate frequency variants contribute much more to the magnitude of $\pi$ than high- and low-frequency variants \cite[eq. 1.4]{Wakeley2009}. Both diversity estimates are significantly higher in \textit{parallelus} than in \textit{erythropus}.
This seems at odds with the expectation from the hitherto proposed demographic model of a postglacial expansion of \textit{C. p. parallelus} towards central and western Europe from a glacial refuge in the Balkans \citep{Cooper1995, Lunt1998}. \textit{C. p. erythropus} in the Pyrenees on the other hand is expected to be derived from several smaller refuges in southern Spain, i. e. its expansion after the last Ice Age would have covered a much smaller distance. According to this model one would expect the \textit{parallelus} subspecies to have undergone a long series of founder events which should have reduced genetic diversity much more so than for the \textit{erythropus} subspecies \citep{Luca2011}.\todo{Look at data from other European species. Do they show the expected reduction of diversity with distance from the the glacial refuge? see \cite{Hewitt2004, Hewitt1996, Hewitt2000, Ramachandran2005, Slatkin2012, Petit1999, Pool2008}}




That ancestral \textit{parallelus} and \textit{erythropus} expanded from their glacial refuges as a "wave", i. e. without colonisation by rare long--distance migration, is made unlikely by at least two observations: 
\begin{itemize}
\item the colonisation of Britain by \textit{C. p. parallelus} before the flooding of the English Channel after the last Ice Age \citep{Cooper1995},
\item the clines for several markers in the Pyrenean hybrid zone between the two subspecies that are too wide to be explained by the \emph{average} dispersal rate of this species of 30 m/generation \citep{Nichols1994}. \todo{\cite{Ibrahim1996}}
\end{itemize}
The higher genetic diversity in Pyrenean \textit{parallelus} is not unique to this study though. So reported \cite{Lunt1998} intraregional $K_{S}$ values of a mitochondrial sequence that were 1.5 times higher in Pyrenean \textit{parallelus} than in Pyrenean \textit{erythropus} and \cite{Llewellyn2008}, table 3--3, found that nucleotide diversity ($\pi_{Tajima}$) was higher in \textit{parallelus} than in \textit{erythropus} from the Pyrenees in 5 out of 7 expressed sequences.\todo{$K_S$ values in \cite{Cooper1995} from a single nuclear locus contradict this and show about twice as much genetic variation in Spain as in France.}


The Tajima's D estimates also differ significantly between the two subspecies (tab. \ref{Tab:diversity_parameters}). Both supspecies have a significantly negative Tajima's D which would be consistent with a recent population expansion.%%
%\begin{figure}[htb]
%\centering
%<< Tajimas-D-by-contig, out.width='.65\\linewidth', echo=FALSE, cache=TRUE, warning=FALSE>>=
%@
%\caption{Distribution of Tajima's D estimates for each contig.}
%\label{Fig:Tajimas-D-by-contig}
%\end{figure}
%%

%
%
%
\section{Demographic history of the hybrid zone}
%
%
%
A simple divergence-in-isolation model in dadi infers a similar divergence time as previously estimated from mitochondrial data (add citation).\todo{add 95\% CI interval to estimated divergence time}. Adding gene flow to this model significantly improves the fit of the simulated model spectrum to the observed spectrum. There has been no divergence in allopatry as previously assumed (add citation). Allowing for gene flow doubles the estimated divergence time. Allowing the gene flow to be asymmetric improves the once more. Gene flow is estimated to have been on average 5 times higher in the direction PAR$\rightarrow$ERY than vice versa. PAR had a much greater ancient effective population size than ERY (see fig. \ref{Fig:overlapping_stairway_plot}). There seems to be a signal of a very recent and drastic bottleneck, at least in PAR. Dadi infers a much less severe bottleneck for ERY than stairway-plot. A recent bottleneck is what would be expected from a series of founder events during range expansion after the last ice age. It would make sense that this had been less severe for ERY than for PAR. A model with ancient gene flow fits the observed 2D SFS better than a secondary contact model. There is no signal of recent gene flow. The detected asymmetry of gene flow can therefore not just be a consequence of different distances to the hybrid zone centre of the two populations or hybrid zone movement. Rather, this hints at intrinsic asymmetric isolation, which is corroborated by asymmetric homogamy in crossings between PAR and ERY (\citep{Bella1992, Hutchison2013}).


\begin{figure}
\includegraphics[width=\textwidth]{ERY_-_200_bootstrap_replicates_final_summary}
\includegraphics[width=\textwidth]{PAR_-_200_bootstrap_replicates_final_summary}
\caption{\texttt{stairway plot} for \textit{erythropus} and \textit{parallelus} from only overlapping sites. The thick grey lines define the 75\% bootstrap-CI and the light grey lines define the 95\% bootstrap-CI.}
\label{Fig:overlapping_stairway_plot}
\end{figure}

\todo[inline, color=green]{check whether gene flow between the two subspecies could explain problems with fitting 1D models, by taking the marginal from the best-fit 2D model that includes migration and determine likelihood}

\todo[inline, color=green]{fit secondary contact model}

\todo[inline, color=green]{include population size change after split into the model}

\todo[inline, color=green]{include unequal migration rate into model}

\todo[inline, color=green]{include isolation after migration model, i. e. only ancient migration model}

\todo[inline, color=green]{apply Ludovic's correction to 1D and 2D unfolded spectra}

\todo[inline, color=green]{get bootstraps for \textbf{un}folded 2D spectrum, bootstrap contigs, not sites as with \texttt{realSFS}}

\todo[inline]{simulate data under best-fit model with ms}

\todo[inline]{investigate deficit of divergently fixed SNP's according to best-fit dadi models}

\todo[inline]{test SFS creation from randomly selected alleles of each genotype in a VCF file (Roger's suggestion to circumvent genotype calling)}

\todo[inline]{Discuss dadi's estimate of divergence time in light of previous estimates: assumptions required for inference, allowing for gene flow, uncertainties}

\todo[inline]{Describe best-fit model and second best-fit model}

\todo[inline]{Learn the theory of creating a genetic map! What kind of data is necessary? Is it conceivable to use genotype likelihoods for that? \cite{Ma2012}}

\todo[inline]{explain inclusion of reads from PCR duplicates. Show plot of coverage distribution from deduplicated read data and non-converged spectra from \texttt{realSFS}.}

\todo[inline]{does dadi's Nref refer to haploid or diploid population size?}

\todo[inline]{Does an outlier test make sense? Would it get us closer to incompatibility loci? According to \cite{Shuker2005} we do not expect alleles at sterility loci to be divergently fixed. So should we look for $F_{ST}$ outliers at all? It might be a nice exercise anyway and we do expect markers that are close to sterility genes to be more divergent than average. The average genome-wide level of $F_{ST}$ is already high.}

\todo[inline]{For the introduction on hybrid sterility, read Dettman's papers on experimental evolution with yeast and Neurospora as well as MacNair's papers on copper tolerant Mimulus.}
%
%
%
\section{Future directions}
\label{sec:future_directions}
%
%
%
Sequencing an individual of a closely related outgroup species, e. g. \textit{Chorthippus montanus}, would allow the polarisation of allele frequencies with respect to ancestral and derived states, assuming that the outgroup sequence generally contains the ancestral state (i. e. that the infinite-sites mutation model holds across the two species). This would allow the analysis of unfolded (i. e. derived allele frequency) spectra and provide greater power to distinguish between alternative demographic models than with the folded (i. e. minor allele frequency) spectra that only could be constructed with the data available for this study \cite[p. 128]{Wakeley2009}.\todo{D-statistics and PBS statistics, Fay and Wu's H statistic}

A moderate increase in the number of sequenced individuals (>50 per population) would ensure enough power to infer more recent demographic events, like a recent secondary contact \citep{Adams2004, Robinson2014}.

In this study, all sequenced grasshoppers were male. There is no male sex chromosome and so all males are haploid for the X chromosome. The X chromosome is the fourth largest of the 9 chromosomes in \textit{Chorthippus parallelus} (8 autosomes plus X) \cite[fig. 2]{Gosalvez1988}. That is, about 1/9th of the RAD loci in this study have been haploid while ANGSD assumed all loci to be diploid. Since these loci appear homozygous they contribute more toward even than uneven frequency classes, thereby creating large deviations from all possible expected counts. This is likely to reduce power to distinguish between demographic models. Since female grasshoppers have two X chromosomes, a future study that attempts to estimate accurate sample allele frequencies should therefore sequence female individuals instead of male individuals. 

It may be possible to find patterns of LD in the data \cite{Kemppainen2015}.


%
%
%
\section{Supplementary Material \& Methods}
\label{sec:suppl_material_methods}
%
%
%
%
%
\subsection{Creation of figure \ref{Fig:sampling-sites-map}}
%
%
The base map was created from two \gls{DEM} tiles  of the \href{http://www.cgiar-csi.org/data/srtm-90m-digital-elevation-database-v4-1}{SRTM 90m Digital Elevation Database v4.1} provided by CGIAR--CSI \citep{Jarvis2008}. I used \href{http://www.qgis.org/en/site/}{QGIS} to extract the polygons corresponding to the country borders of France and Spain from a shapefile containing all world-wide country borders (downloaded from \href{http://www.naturalearthdata.com/downloads/10m-cultural-vectors/10m-admin-0-countries/}{Natural Earth}).
Raster manipulations (projection, merging, clipping to polygon outline, color relief, hill shade and slope shade)  were carried out with utilities from the \href{www.gdal.org}{Geospatial Data Abstraction Library (GDAL)}. The exact GDAL command lines can be provided upon request.  Raster overlays and map design was performed with \href{https://tilemill-project.github.io/tilemill/}{\texttt{TileMill}}, which renders maps using \href{http://mapnik.org/}{\texttt{mapnik}}. Final map image manipulations were performed with \href{https://inkscape.org/en/}{InkScape}. 

%
%
%
\subsection{How well are mapping quality scores calibrated?}
%
%
%
The mapping quality score (mapQ) is the Phred scaled probability $p$ that the true mapping position of the read is not the reported mapping position (rounded to the next integer):
$$
\text{mapQ} = -10log_{10}p
$$
%
\textsl{True mapping position} is meant with respect to the provided \textsl{reference sequence}, i. e. assuming the true reference sequence is known without error. 
%
%\begin{figure}[htb]
%\raggedright
%a)
%\includegraphics[width=\textwidth]{obs-vs-reported-FPR-1}
%b)
%\includegraphics[width=\textwidth]{obs-vs-reported-FPR-Hmel-Hsap-1}
%\caption{Comparison of reported and observed false positive mapping rate (FPR). The reported FPR is the mapping quality score transformed back into a probability. The observed FPR is the proportion of all mapped reads with a specific mapping quality score that have an incorrect mapping position reported by the mapping programme. The diagonal line indicates 1:1 correspondence. a) Simulating Illumina reads from the \textit{Heliconius melpomene} reference sequence and mapping with different mapping programmes. b) Simulating Illumina reads from \textit{Heliconius melpomene} and \textit{Homo sapiens} reference sequence and mapping with \texttt{bowtie2}.
%For points above the diagonal the mapping programme underestimates the true mapping quality. For \texttt{bowtie2} several points have been annotated with the reported mapping quality score of the reads they represent.}
%\label{Fig:obs-vs-reported-FPR}
%\end{figure}
%
\begin{figure}
\centering
\subcaptionbox{Simulating Illumina reads from the \textit{Heliconius melpomene} reference sequence and mapping with different mapping programmes.\label{Fig:obs-vs-reported-FPR-a}}[\textwidth]
	{\includegraphics[width=\textwidth]{obs-vs-reported-FPR-1}}
\hfill
\subcaptionbox{Simulating Illumina reads from \textit{Heliconius melpomene} and \textit{Homo sapiens} reference sequence and mapping with \texttt{bowtie2}.\label{Fig:obs-vs-reported-FPR-b}}[\textwidth]
	{\includegraphics[width=\textwidth]{obs-vs-reported-FPR-Hmel-Hsap-1}}
\caption{Comparison of reported and observed false positive mapping rate (FPR). The reported FPR is the mapping quality score transformed back into a probability. The observed FPR is the proportion of all mapped reads with a specific mapping quality score that have an incorrect mapping position reported by the mapping programme. The diagonal line indicates 1:1 correspondence. For points above the diagonal the mapping programme underestimates the true mapping quality. For \texttt{bowtie2} several points have been annotated with the reported mapping quality score of the reads they represent.}
\label{Fig:obs-vs-reported-FPR}
\end{figure}
%
I have simulated 200,000 \gls{se} reads of 50 bp length from the \textit{Heliconius melpomene} genome reference (Hmel1-1\_Release\_20120601) with the programme \texttt{mason} (version 0.1.2) \citep{Holtgrewe2010}. The \textit{H. melpomene} reference sequence has a length of 269,658,870 bp. I have also simulated reads from the much bigger (and probably more repetitive) human genome reference sequence (version GRCh38). The human reference sequence is 11.5 times larger than the \textit{H. melpomene} reference sequence (3,099,750,718 bp).
I simulated reads without sequencing error but with a polymorphism rate of 0.01, 20\% of which were \glspl{indel}. Indels could be up to 20 bp long and their length was drawn from a uniform distribution. 
\texttt{bowtie2} was run in \texttt{--very-sensitive} and \texttt{--end-to-end} mode, \texttt{bwa mem} and \texttt{stampy} were run with default settings. In order to extract counts of mapped and incorrectly mapped reads from SAM files, I used the programme \texttt{wgsim\_eval.pl} from the read simulation package \href{https://github.com/lh3/wgsim}{\texttt{wgsim}} of Heng Li. A read is mapped correctly if its reported mapping position is within 50 bp of the true origin. Further details can be found in \href{https://github.com/claudiuskerth/PhDthesis/blob/master/Data_analysis/reference-mapping/data/simulation/Mason/Mason_sim.sh}{Mason\_sim.sh} and \href{https://github.com/claudiuskerth/PhDthesis/blob/master/Data_analysis/reference-mapping/data/simulation/Mapping_Tool_Test.md#how-well-are-mapq-scores-calibrated}{\texttt{Mapping\_Tool\_Test}}.

The observed rate of mismappings of these simulated reads should be an estimator of the true mapping quality $p$. I can therefore compare how well the reported mapping qualities of the different mapping programmes correspond to the observed false positive rate. Figure \ref{Fig:obs-vs-reported-FPR} shows that \texttt{bowtie2}, generally underestimates the true mapping quality, given the reference sequence, when simulating from the \textit{H. melpomene} reference. When mapping against the more repetitive human reference \texttt{bowtie2} underestimates mapping quality for quality scores below 25 and slightly overestimates mapping quality for reads with higher quality scores assigned.

%
%
%
\subsection{Mappability across the de novo reference assembly}
\label{Sec:mappability}
%
%
%
Mappability is a measures that can be computed for each single nucleotide position $x$ in a reference sequence, assuming that the reference sequence is known without error. Starting at position $x$ a sequence of length $k$ is extracted from the reference sequence. This \gls{kmer} is then searched for in the whole reference sequence allowing for up to $m$ mismatches (or edits, i. e. including \glspl{indel}). The number of times the kmer is found in the reference is $F_{k,m}$. Mappability $M$ is then defined as:
$$
M_{k,m}(x) = \frac{1}{F_{k,m}(x)}
$$
I have used programmes in the tool package \href{http://algorithms.cnag.cat/wiki/The_GEM_library}{GEM} \citep{Derrien2012} together with programmes from the tool package \texttt{BEDOPS} \citep{Neph2012} to compute the average mappability over the positions in each contig (see \texttt{assembly.sh} for further details). I specified a kmer length of 40 and allowed up to 3 edits during the search for mapping positions in the Big Data reference assembly. The search did not include the reverse complement of the reference. Average mappability per contig was computed for all contigs in the complete unfiltered Big Data reference assembly and the subset of contigs which passed all filters, i. e. which included at least one interval after filtering that was kept for downstream analysis. The difference in location between the distributions in figure \ref{Fig:mappability-dist} was tested with the \texttt{R} function \texttt{wilcox.test}. 

% ----------------------------

\begin{figure}
\includegraphics[width=\textwidth]{ERY_-_200_bootstrap_replicates_final_summary_nonO}
\includegraphics[width=\textwidth]{PAR_-_200_bootstrap_replicates_final_summary_nonO}
\caption{\texttt{stairway plot} for \textit{erythropus} and \textit{parallelus} from including non-overlapping sites. The thick grey lines define the 75\% bootstrap-CI and the light grey lines define the 95\% bootstrap-CI.}
\label{Fig:nonO_stairway_plot}
\end{figure}

%
%
\section{TODO}
%
%
%
% add [$\surd$] after \item to mark entry as completed
% \ding{} requires \usepackage{pifont}
\begin{itemize}
\item Rewrite introduction:
\begin{itemize}
\item literature about the hybrid zone
\item ice ages, climatic history
\item other phylogeographic and demographic studies
\item other hybrid zones
\item evolution of hybrid incompatibilities (Dettman and Mimulus)
\end{itemize}
\item Material and Methods:
\begin{itemize}
\item[\ding{52}] added standard RAD protocol to appendix of thesis
\item[\ding{52}] moved description of map creation to suppl. mat.
\item[\ding{52}] linked to \texttt{assembly.sh} at github (not bitbucket)
\item[\ding{52}] included de novo assembly sketch
\item[\ding{52}] produced a distribution of mapping qualities before and after filtering
\item[\ding{52}] explained mapping quality of 5 in \texttt{bowtie2}, see \texttt{Mapping\_Tool\_Test.Rmd}
\item[\ding{52}] included a plot of the length distribution of contigs after filtering
\item include plot of distribution of average mappability per contig before and after filtering
\item specify ANGSD version for HWE filtering, mention the difference in BAQ to the version used later
\item summarize and document positive Fis estimates (see \texttt{assembly.sh} and \texttt{MAF\_by\_pval.ipynb})
\item add deduplication analysis, show coverage distribution
\item redo PCA with new version of ANGSD and extended BAQ
\item redo Fst analysis with new version of ANGSD and extended BAQ
\item document bootstrapping over contigs in genetic diversity section
\item document determination of overlapping sites and estimation of statistics from those
\item add subsection brief description of dadi analysis with links to IPython notebooks, should include a brief description of how it works
\item add subsection with brief description of stairway-plot with custom version, briefly describe how it works 
\end{itemize}
\item Results:
\begin{itemize}
\item document Fst analysis, bootstrapping, Fst.Rmd
\item estimate Fst from 2D SFS with dadi, including bootstraps
\item replace 2D SFS with one that contains logarithm of counts (use dadi for that)
\item read about the potential effects of allele-dropout on SFS and discuss
\item read about potential reasons for negative Tajima's D and discuss
\item read up on serial founder effect and discuss
\item mention that 1D model fitting with dadi was not successful
\item add CI to divergence time in simple divergence model
\item add CI to divergence time in divergence with gene flow model
\item add plot for spectra with varying divergence time and varying migration rate
\item aren't divergence time and migration rate conflated? how can dadi distinguish between those?
\item was gene flow low enough to allow divergence by drift (see \cite{Bank2012})
\item add figure of 2-epoch model
\item add figure of primary contact model
\item add table with model parameters for the 2-epoch and primary contact model including 95\% CI
\item What happened $\sim$1.5 Mya in the Pyrenees? \cite{Webb1992, Hewitt2011, Hewitt2004, Sommer2009}
\item discuss assumed mutation rate and the influence my uncertainty in it on parameter estimates in absolute units, \cite{Liu2017, Keightley2009}
\end{itemize}
\end{itemize}

%----------------------------
% Bibliography
%----------------------------

\bibliographystyle{elsarticle-harv}
\bibliography{/Users/Claudius/Documents/MyLiterature/Literature}
%

%----------------------------
% Appendices
%----------------------------
%\begin{appendices} % Using appendices environment for more functunality
\chapter{Appendix}

\input{../Appendix1/appendix1}


%\end{appendices}

\end{document}

