2DSFS
================
Claudius
24/11/2016

-   [Which population on which axis?](#which-population-on-which-axis)
-   [Population differentiation](#population-differentiation)
-   [Diversity Estimates](#diversity-estimates)
    -   [Tajima's D](#tajimas-d)
    -   [Nucleotide Diversity](#nucleotide-diversity)

I have created a *folded* 2D-SFS with `ANGSD` and `realSFS`.

``` r
# realSFS has produced a flattened matrix with format nrow(pop2) x ncol(pop1)
sfs2d = scan("ERY_PAR.2dsfs")
# a matrix is filled column wise:
matrix(1:16, nrow=4, ncol=4)
```

    ##      [,1] [,2] [,3] [,4]
    ## [1,]    1    5    9   13
    ## [2,]    2    6   10   14
    ## [3,]    3    7   11   15
    ## [4,]    4    8   12   16

``` r
sfs2d = matrix(sfs2d, nrow=19, ncol=19) # rows should be PAR, columns should be ERY
sfs2d
```

    ##               [,1]        [,2]        [,3]        [,4]        [,5]
    ##  [1,] 918718.73720 3306.810792 2658.501012 1254.230913 1125.282582
    ##  [2,]   5436.18750  334.314623  176.801605  212.279890   56.384101
    ##  [3,]   6006.04886  327.699176  216.840553  143.659315    0.420384
    ##  [4,]   2257.59768  110.979093   50.013173  167.165704   53.174357
    ##  [5,]   1165.29024   87.641436  171.314167    4.866786  109.105148
    ##  [6,]   1098.18039    0.174917   23.649713   23.633598   40.273671
    ##  [7,]    634.60532   91.681510  105.578977   51.490235    3.579671
    ##  [8,]     95.25376    0.247015   25.832276   31.976818    8.116351
    ##  [9,]    525.61764    0.652448    0.002414   37.008974  149.378507
    ## [10,]    361.83689  120.355093    0.002636   35.503963   56.526703
    ## [11,]     62.72300    1.800923   15.246331    7.912195   12.293012
    ## [12,]    183.66582   34.429541    0.145540    0.003132    3.115636
    ## [13,]    360.88646   62.721313   83.945580    2.829655    4.399697
    ## [14,]    120.13998    2.828493   53.598841   45.929963    3.089096
    ## [15,]     37.17949    2.022539    2.477976   18.374727   10.427798
    ## [16,]    215.91736    8.006507    1.925712   28.126336    0.030271
    ## [17,]     12.13773   12.007092    1.166763   12.034147    0.000468
    ## [18,]    136.58046   15.016088   34.884290   20.108390    0.000000
    ## [19,]    160.16426    8.140144   14.817615   29.891874    0.000000
    ##             [,6]       [,7]       [,8]       [,9]      [,10]      [,11]
    ##  [1,] 873.768656 584.023778 275.388786 616.941997  95.058874 308.004710
    ##  [2,]  19.188160  22.109546  67.078179  15.072787   0.028007   0.000000
    ##  [3,] 178.234700  42.649946   0.028967   0.510581 155.554059   0.001085
    ##  [4,] 133.690783   2.991422  76.226948  40.726848  39.698465   0.000000
    ##  [5,]  14.807418  17.271089  27.065199  48.677708  89.031730   2.626076
    ##  [6,] 159.506161   0.635899  39.585381  14.928274   2.237695   0.000034
    ##  [7,]   0.105361   0.000000 107.060870  61.909050   5.127169   0.059214
    ##  [8,]   0.278025   0.001575   0.877099  18.311226   2.473746  59.851831
    ##  [9,]  49.680258   4.407956   4.981107   1.895496   0.103488  25.790185
    ## [10,]  16.822280  74.093193   1.522087  95.106416   0.000000   0.019717
    ## [11,]   2.681995  78.437113   4.531314  82.672527  36.778580   8.156307
    ## [12,]   0.029005   1.560741   2.374534   3.341600   1.104933   0.463707
    ## [13,]   1.659532   0.833901   2.645267 132.521591   0.264182   0.122999
    ## [14,]  25.050961   1.602351  26.189830   2.938216   0.245197   0.978830
    ## [15,]  97.855182   3.911094  20.081730   2.406509   0.016857   6.426442
    ## [16,]  20.972548  11.122469  16.947643   0.269405   0.519192  27.095770
    ## [17,]   0.000000   8.415022  15.453988   5.999398   0.072999  51.116183
    ## [18,]   0.000317   2.095199  11.014329   4.828438   0.002972   0.137724
    ## [19,]   0.000736   4.353217  16.338024  66.241053   0.000264   0.000000
    ##            [,12]      [,13]      [,14]      [,15]      [,16]      [,17]
    ##  [1,] 311.158461 344.431373 129.211196 159.621879 109.787504 328.306290
    ##  [2,]  26.974089  20.159906  10.949741  56.923516   0.000870   0.009271
    ##  [3,]  12.555989  63.375859   0.541325  11.402476   5.646113  86.572596
    ##  [4,]   7.099276  58.196032  57.323664   0.000000   0.122052  16.646163
    ##  [5,]   0.000000   0.000000   2.594219   6.973973   1.186274   0.040000
    ##  [6,]   0.659042  59.564986   4.575973 133.533847  12.446152  11.243037
    ##  [7,]  14.567411  10.699298   0.971674   0.848476   0.401150   1.395258
    ##  [8,]   0.263355  59.721488   0.327913   0.000000   0.000604   0.029719
    ##  [9,]   0.007771 137.377364   0.295243   0.003935   0.000000   0.496345
    ## [10,]   0.002684  32.192400   1.029732   0.006667   5.442608  42.599727
    ## [11,]   0.000552   0.051014   0.145079   0.004977 126.049657  32.663339
    ## [12,]   0.005185   0.038839   0.223840   0.025564   5.931971  16.100995
    ## [13,]   0.630521   3.603966  17.606597   0.870590   2.873659  26.028102
    ## [14,]  51.440392  16.282977 127.951328   7.779625  10.130388  32.551910
    ## [15,]  51.098417   1.120572  36.395030  10.402144   8.018998  33.642106
    ## [16,]   2.221780   0.269538  12.610454   0.102931   2.302604  22.549530
    ## [17,]  90.955488   0.703094   4.646852   0.032259   8.101337  14.001367
    ## [18,]   2.346336   0.055732   2.799269   1.909763  10.207233   2.275543
    ## [19,]   0.023723   0.004762  10.157307   0.662195  76.797326  14.974812
    ##           [,18]      [,19]
    ##  [1,] 27.563580 180.441880
    ##  [2,]  7.072328   4.600172
    ##  [3,] 11.251670  19.131115
    ##  [4,]  1.717986  42.026419
    ##  [5,]  0.051489  23.721530
    ##  [6,]  0.019360   5.002960
    ##  [7,]  0.292098 101.748496
    ##  [8,]  0.005743   2.920919
    ##  [9,]  0.061494  10.772749
    ## [10,]  2.827850   7.802636
    ## [11,]  3.636332   0.521688
    ## [12,] 10.666156   6.530408
    ## [13,]  3.099161   2.670597
    ## [14,]  2.993790   0.158756
    ## [15,]  2.398253   0.837985
    ## [16,]  3.168918  15.348498
    ## [17,]  3.226456  44.480282
    ## [18,]  0.101837  21.010629
    ## [19,]  0.926251  13.942763

``` r
sfs2d[1,1] = 0
image(0:18, 0:18, sfs2d, xlab="PAR", ylab="ERY", main="2D-SFS", col=heat.colors(22))
```

![](2DSFS_files/figure-markdown_github/unnamed-chunk-1-1.png)

``` r
library(fields)
```

    ## Loading required package: spam

    ## Loading required package: grid

    ## Spam version 1.4-0 (2016-08-29) is loaded.
    ## Type 'help( Spam)' or 'demo( spam)' for a short introduction 
    ## and overview of this package.
    ## Help for individual functions is also obtained by adding the
    ## suffix '.spam' to the function name, e.g. 'help( chol.spam)'.

    ## 
    ## Attaching package: 'spam'

    ## The following objects are masked from 'package:base':
    ## 
    ##     backsolve, forwardsolve

    ## Loading required package: maps

``` r
pdf("ERY_PAR.2dsfs.pdf")
# rows in the matrix are on the x-axis, columns are on the y-axis:
image.plot(0:18, 0:18, sfs2d, xlab="PAR", ylab="ERY", main="folded 2D-SFS")
dev.off()
```

    ## pdf 
    ##   2

``` r
sfs2d_prop = sfs2d/sum(sfs2d)
image.plot(0:18, 0:18, sfs2d_prop, xlab="PAR", ylab="ERY", main="folded 2D-SFS")
```

![](2DSFS_files/figure-markdown_github/unnamed-chunk-3-1.png)

### Which population on which axis?

I have created an undolded 2D-SFS with `ANGSD` and `realSFS`, but this time giving it only 10 individuals for population ERY. I gave to `realSFS` the two *saf.idx* files, the one from ERY before the one from PAR. So there are 21 categories for ERY and 37 for parallelus (2\*n+1).

The 2D matrix is output in a flattened format by `realSFS`. After running `tr ' ' '\n' < *2dsfs | less -N`, I convinced myself that every 38th number is much higher than every 37th number. This indicates that the flattened matrix contains the concatenated columns (not rows) of the 2D matrix and that PAR is on the *row* dimension and ERY on the *column* dimension.

The help text for the function `image.plot` from the package `fields` says:

> If the z component is a matrix then the user should be aware that this function locates the matrix element z\[i,j\] at the grid locations (x\[i\], y\[j\])

That means that the rows of the matrix are plotted on the x-axis (PAR) and the columns on the y-axis (ERY).

``` r
library(fields)
setwd("/data3/claudius/Big_Data/ANGSD/2D-SFS-axes")
sfs2d = matrix(scan("ERY.PAR.unfolded.2dsfs"), nrow=37, ncol=21)
sfs2d[1,1] = 0
image.plot(0:36, 0:20, sfs2d, xlab="ERY", ylab="PAR", main="unfolded 2D-SFS")
```

![](2DSFS_files/figure-markdown_github/unnamed-chunk-4-1.png)

``` r
sfs2d[,1:15]
```

    ##              [,1]       [,2]       [,3]       [,4]       [,5]      [,6]
    ##  [1,]    0.000000 819.566137 456.815004 257.475033 139.406744 83.668274
    ##  [2,] 1911.512200  97.713058  49.447018  10.435684   4.661277  9.704523
    ##  [3,] 1068.136304  79.333343  41.976616   0.347967  29.536802  6.696707
    ##  [4,]  593.456321  38.387358  42.594066  57.775536   2.220740  0.000063
    ##  [5,]  278.326399  16.351674  20.261597   4.334924   0.000655 13.848433
    ##  [6,]  223.935827   9.650408   2.931154  34.741923   0.000000  5.359546
    ##  [7,]   94.643441   0.005362  19.532607   0.003217   0.000196  0.000000
    ##  [8,]   79.075609  34.340680  27.008144   0.005556  27.449810  8.279875
    ##  [9,]  122.242104   0.000000   9.893726  11.236117   7.811002  0.109658
    ## [10,]    4.813906  15.336778   0.214486   0.000000   0.049708  1.884522
    ## [11,]   40.973730   0.022805   0.020574   0.000000   5.960970  0.274039
    ## [12,]   26.919629  18.477111   9.476754   0.000003  20.749090  0.062785
    ## [13,]   70.050339   0.001349  18.173042  19.745256   0.000000  2.198449
    ## [14,]    1.463535   0.000003   0.008572   0.003898   0.000000  1.148723
    ## [15,]    4.213883   0.000003   0.011954   0.003586   0.000002  0.001820
    ## [16,]   27.578325   0.003678  14.515829   0.000154   0.000000 24.492641
    ## [17,]   34.429552   0.369692   3.310520   0.011087   0.000001  8.529873
    ## [18,]   11.544450  13.580419   0.000000   0.085948   0.000003  0.029905
    ## [19,]    1.644831   0.000000   0.000383   9.218254   0.002374  1.172950
    ## [20,]   15.249262   0.001400   0.000284   0.000000   0.307325 11.521419
    ## [21,]   20.869172   0.000000   0.000000   0.000000   2.770027  0.011540
    ## [22,]   18.370778   0.000014   6.343783   0.000001   2.463423  0.000000
    ## [23,]    8.948196   0.000000   2.913178   0.000000   4.408533  0.000005
    ## [24,]    6.693984   0.000000   0.002882   0.000000   0.000000  0.000020
    ## [25,]    0.020550   0.000000   0.008578   0.000000   0.000000  0.000000
    ## [26,]    1.378799   0.000000   0.021461   0.000000   0.000000 11.655560
    ## [27,]   39.583486   0.000000   8.741389   0.000000   0.000000  0.102349
    ## [28,]    0.000000   0.000012   0.000280   0.003055   0.000001  0.000039
    ## [29,]    0.000000   0.330220   0.000000   7.845423   0.001388  0.000001
    ## [30,]    0.000000   2.867615   0.000000   0.001052   3.882047  0.000866
    ## [31,]    0.022591   0.000000   0.000000   0.101276   0.000000  5.929500
    ## [32,]   19.456510   0.000000   0.000007   4.465049   0.000000  0.011829
    ## [33,]    1.089870   0.000000  10.675842   0.000783   0.000000  0.000000
    ## [34,]    8.073310   0.000000   0.000000   0.000000   0.000000  0.000000
    ## [35,]    0.000000   0.000000   0.000000   0.000000   0.000000  0.000000
    ## [36,]    5.274045   0.000000   3.093718   5.036013   0.000000  0.000000
    ## [37,]   34.998233   3.365538   6.742485   0.038844   4.942237  1.908484
    ##            [,7]      [,8]      [,9]     [,10]     [,11]     [,12]
    ##  [1,] 50.495983 78.157856 22.072331  9.535526 33.641582 20.079289
    ##  [2,]  9.438042  4.589135  0.000000  0.608525 10.996733  0.036555
    ##  [3,] 14.863727  0.004126  0.001097  6.268875  1.027786  0.000761
    ##  [4,]  5.259418  0.025564 18.798071  1.230744  3.325348 12.863545
    ##  [5,]  0.000960  0.001324  0.005541  5.279667  3.782259  0.000000
    ##  [6,]  0.059180 13.217778  0.000000  0.000001  0.000000  0.000000
    ##  [7,] 12.053465  1.841200  0.064571  0.000000  0.012711  0.000000
    ##  [8,]  4.883696 15.029503  1.075517  9.147285 11.289729  0.000000
    ##  [9,]  0.356659  0.564023  0.000000  7.159937  0.005057  3.559632
    ## [10,]  2.190274  9.434328  7.253843  0.000002  0.000000  0.000000
    ## [11,] 27.218291  0.025613  0.020771  0.000000  0.000000  0.000006
    ## [12,]  0.000000  0.000167  0.002283  0.000010  0.000000  0.000000
    ## [13,]  0.330633  0.003992  0.001334  1.677610  0.000000  0.000000
    ## [14,]  6.211883  0.193601 10.215074 10.621408  0.000000  0.000000
    ## [15,]  0.009681  5.620880  0.000000  7.596155  0.000000  8.207860
    ## [16,]  0.000347 14.068507  0.000031  0.115073  0.035164  0.003416
    ## [17,]  0.003172  0.176072  0.000001  0.205673  4.488016  0.000001
    ## [18,]  0.002031  0.012713  0.000249  1.338457  2.710259  0.000143
    ## [19,]  0.000000  0.005272  0.001507  1.943062  5.016931 12.633550
    ## [20,]  3.111549 14.325847  9.028146  0.775931  0.488942  5.535464
    ## [21,]  0.009802  0.045249  0.133864  0.000000  3.816219  0.000003
    ## [22,]  0.000000  0.000009  0.002156  0.000185  0.856815  0.000000
    ## [23,]  0.000000  0.000000  0.000260  0.000001  1.668210  0.000000
    ## [24,]  0.000000  0.000000  0.001362  0.000001 17.829995  0.000000
    ## [25,]  0.000000  0.000000  4.863318  0.000003  5.187410  0.000025
    ## [26,]  0.000001  0.000000  0.106028  0.000002  0.011386  0.004814
    ## [27,]  9.725579  0.000001  9.807769  0.000000  0.000000  0.003915
    ## [28,]  0.037487  0.000000  2.361090  0.000001  0.000000  0.001320
    ## [29,]  0.000099  0.000000  0.000001  0.000006  0.000144  0.001397
    ## [30,]  0.001966  0.000124  0.000000  0.000000  0.000000  0.000181
    ## [31,]  7.784657  5.669382  0.000000  4.289705 15.036330  0.000000
    ## [32,]  0.011057  0.000000  0.000000  2.052464  0.039082  0.000000
    ## [33,]  0.003485  1.391984  0.000000  0.020674  0.000013  0.000000
    ## [34,]  3.608035  0.095371  0.004358  8.514858  0.000000  0.000000
    ## [35,]  2.458280  0.141243  0.223380  0.000372  0.000000  0.000000
    ## [36,]  2.252120  4.570966  0.000855  0.000612  3.893783  0.005808
    ## [37,]  2.410330 10.691952  0.000000  2.558897  9.307333 17.391360
    ##           [,13]     [,14]     [,15]
    ##  [1,]  3.964944 10.719747  4.411819
    ##  [2,]  0.000000  0.000000  1.024731
    ##  [3,]  0.000049  4.530408  7.235883
    ##  [4,]  0.000002  1.267085  5.353423
    ##  [5,]  0.000000  0.000000  0.000000
    ##  [6,]  0.000000  0.000000  0.000000
    ##  [7,]  0.000000  0.000028  0.000001
    ##  [8,]  0.000000  7.231649  0.000000
    ##  [9,]  0.000000  0.054795  0.000000
    ## [10,]  0.000005  0.001272  0.000000
    ## [11,] 12.214373  0.000425  0.000000
    ## [12,]  0.131133  0.000017  0.000000
    ## [13,]  0.000003  0.000010  0.000000
    ## [14,]  0.000462  0.000226  0.000881
    ## [15,]  0.001630 17.267397  0.323641
    ## [16,]  0.001343  0.127817  0.001930
    ## [17,]  0.000785  0.000656  0.000001
    ## [18,]  0.000000  0.000000  0.000000
    ## [19,]  0.853761  0.000057  0.000001
    ## [20,]  1.106874  0.000000  0.002040
    ## [21,]  0.000009  0.009805 20.846157
    ## [22,]  0.000028  0.001182  0.741799
    ## [23,]  0.000000  0.000426  0.002659
    ## [24,]  0.000001  0.443915  0.008368
    ## [25,]  0.001197 26.149634  0.013708
    ## [26,]  0.079926  0.066700  0.153388
    ## [27,]  0.007017  0.013023 13.974624
    ## [28,]  0.002572  0.000035  0.398590
    ## [29,]  0.151101  0.000226  3.961673
    ## [30,]  8.621364  0.000004  0.000000
    ## [31,]  0.104402  0.000000  0.000000
    ## [32,]  0.004644  8.287672  0.000000
    ## [33,]  0.000000  6.662498  0.000000
    ## [34,]  0.000000  0.000000  0.000002
    ## [35,]  0.000040  0.000000 17.796562
    ## [36,]  8.165826  4.532492  0.000000
    ## [37,]  6.913231  4.457843 14.335709

Population differentiation
--------------------------

I have estimated unfolded SAF's for Par and Ery populations separately and then used these to estimate a ML 2D-SFS. I then used this 2D-SFS as prior for estimating Fst (see assembly.sh and Fumagalli2013).

``` r
fst = read.delim("ERY.PAR.fst.tab", header=TRUE)
par(mfrow=c(1,1))
hist(fst[,3]/fst[,4], col="grey", xlab="Fst", main="Fst distribution")
```

![](2DSFS_files/figure-markdown_github/FST-distribution-1.png)

As can be seen from this distribution plot, the vast majority of *F*<sub>*S**T*</sub> values is close to 0. But note, that *F*<sub>*S**T*</sub> was calculated for all positions, i. e. including those which are likely monomorphic. Alos, many positions are linked. An alternative could to extract only the highest *F*<sub>*S**T*</sub> from each contig.

``` r
A = fst[,3] # between population genetic variance
A_plus_B= fst[,4] # total genetic variance
B = A_plus_B - A
FST = A / A_plus_B # equation (3) in Fumagalli2013
FST[which(A_plus_B==0)] = 0 # replace NAN caused by division by zero in the previous command with 0
any(A_plus_B < 0)
```

    ## [1] FALSE

``` r
length(fst$Chrom) # number of positions with Fst estimate
```

    ## [1] 1561867

``` r
# png("ERY.PAR.fst.png") # I am going to plot >1M data points, don't want to store all of them in a vector graphics file, hence pixel graphics file
# par(mfrow=c(1,1))
# # this may take a few seconds:
# plot(x=A_plus_B, y=FST, xlim=c(0,1), ylim=c(0,1),pch=20, col=grey(0, 0.2), xlab="a+b", ylab="a/(a+b)")
# dev.off()
```

There are 0 positions for which I have a ML estimate of *F*<sub>*S**T*</sub>. The genome wide *F*<sub>*S**T*</sub> is 0.294272. This is a **weighted** *F*<sub>*S**T*</sub>, i. e. at each position sample sizes are used to weigh variance estimates.

Diversity Estimates
-------------------

### Tajima's D

``` r
thetasPAR = read.table("PAR.thetas.tab", header=TRUE)
head(thetasPAR)
```

    ##          Chr       tW       tP    Tajima nSites
    ## 1  Contig_40 0.706850 0.279552 -1.572534     39
    ## 2  Contig_53 0.255613 0.105083 -1.053296     37
    ## 3 Contig_179 0.210184 0.158631 -0.404111     43
    ## 4 Contig_195 0.304229 0.188745 -0.728733     38
    ## 5 Contig_219 0.428563 0.161175 -1.366711     39
    ## 6 Contig_252 0.263451 0.128004 -0.931071     39

``` r
thetasERY = read.table("ERY.thetas.tab", header=TRUE)
head(thetasERY)
```

    ##         Chr       tW       tP    Tajima nSites
    ## 1 Contig_16 0.013960 0.003707 -0.335888     39
    ## 2 Contig_17 0.011645 0.003195 -0.303390     39
    ## 3 Contig_40 0.489737 0.428333 -0.288277     38
    ## 4 Contig_46 0.498307 0.831142  1.545205     39
    ## 5 Contig_53 0.015003 0.004305 -0.337916     39
    ## 6 Contig_99 0.257062 0.145092 -0.780889     35

``` r
# per-contig Tajima's D distribution:
#pdf("TajimasD-dist.pdf")
par(mfrow=c(2,1))
hist(thetasPAR$Tajima, xlab="Tajima's D", main="parallelus", col="grey")
hist(thetasERY$Tajima, xlab="Tajima's D", col="grey", main="erythropus")
```

![](2DSFS_files/figure-markdown_github/Tajimas-D-dist-1.png)

``` r
#dev.off()
```

### Nucleotide Diversity

"tW" is Watterson's *θ*, i. e. number of segregating sites. "tP" is *π*, but per contig. So I have to divide that by the number of sites in the contig to get the per-nt *π* estimate.

``` r
piPAR = thetasPAR$tP/thetasPAR$nSites
piERY = thetasERY$tP/thetasERY$nSites
thetasPAR[which(is.na(piPAR)),]
```

    ##                 Chr tW tP Tajima nSites
    ## 2443   Contig_55775  0  0      0      0
    ## 3574   Contig_78912  0  0      0      0
    ## 6180  Contig_135061  0  0      0      0
    ## 6264  Contig_137431  0  0      0      0
    ## 7161  Contig_158524  0  0      0      0
    ## 17402 Contig_407197  0  0      0      0
    ## 19573 Contig_494767  0  0      0      0
    ## 20600 Contig_572196  0  0      0      0
    ## 20663 Contig_576782  0  0      0      0

``` r
thetasERY[which(is.na(piERY)),]
```

    ##                 Chr tW tP Tajima nSites
    ## 5008   Contig_78912  0  0      0      0
    ## 8664  Contig_135061  0  0      0      0
    ## 10172 Contig_158524  0  0      0      0
    ## 26578 Contig_411500  0  0      0      0
    ## 27454 Contig_431207  0  0      0      0
    ## 30913 Contig_572196  0  0      0      0
    ## 31920 Contig_672702  0  0      0      0

``` r
piPAR[which(is.na(piPAR))] = 0
piERY[which(is.na(piERY))] = 0
length(piPAR)
```

    ## [1] 21811

``` r
length(piERY)
```

    ## [1] 32252

``` r
mean(piPAR)
```

    ## [1] 0.005682508

``` r
mean(piERY)
```

    ## [1] 0.005206228

``` r
# distribution of nucleotide heterozygosity:
#pdf("NucleotideDiv.pdf")
par(mfrow=c(2,1))
hist(piPAR, breaks=seq(0, .06, .005), col="gray", xlab=expression(pi[nt]), main="parallelus")
text(0.03, 6000, labels=expression(paste("average ", pi[nt] == 0.00568) ) )
# ERY seems to have some contigs with extreme pi, i. e. >0.06:
hist(piERY[piERY<.06], breaks=seq(0, .06, .005), col="gray", xlim=c(0, 0.06), xlab=expression(pi[nt]), main="erythropus")
text(0.03, 6000, , labels=expression(paste("average ", pi[nt] == 0.00521) ) )
```

![](2DSFS_files/figure-markdown_github/dist-nucl-het-1.png)

``` r
#dev.off()
```
