<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Claudius" />

<meta name="date" content="2016-12-12" />

<title>PCA</title>

<script src="PCA_files/jquery-1.11.0/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="PCA_files/bootstrap-3.3.1/css/cosmo.min.css" rel="stylesheet" />
<script src="PCA_files/bootstrap-3.3.1/js/bootstrap.min.js"></script>
<script src="PCA_files/bootstrap-3.3.1/shim/html5shiv.min.js"></script>
<script src="PCA_files/bootstrap-3.3.1/shim/respond.min.js"></script>

<style type="text/css">code{white-space: pre;}</style>
<link rel="stylesheet"
      href="PCA_files/highlight/default.css"
      type="text/css" />
<script src="PCA_files/highlight/highlight.js"></script>
<style type="text/css">
  pre:not([class]) {
    background-color: white;
  }
</style>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>



</head>

<body>

<style type = "text/css">
.main-container {
  max-width: 940px;
  margin-left: auto;
  margin-right: auto;
}
code {
  color: inherit;
  background-color: rgba(0, 0, 0, 0.04);
}
img { 
  max-width:100%; 
  height: auto; 
}
</style>
<div class="container-fluid main-container">


<div id="header">
<h1 class="title">PCA</h1>
<h4 class="author"><em>Claudius</em></h4>
<h4 class="date"><em>12/12/2016</em></h4>
</div>

<div id="TOC">
<ul>
<li><a href="#pca-with-known-minor-allele-and-snp-calling">PCA with known minor allele and SNP calling</a></li>
<li><a href="#pca-with-unknown-minor-allele-and-snp-calling">PCA with unknown minor allele and SNP calling</a></li>
<li><a href="#pca-with-unknown-minor-allele-snp-calling-genotype-calling">PCA with unknown minor allele + SNP calling + genotype calling</a></li>
<li><a href="#pca-with-unknown-minor-allele-and-without-snp-calling">PCA with unknown minor allele and without SNP calling</a></li>
<li><a href="#ery-and-par-separately">ERY and PAR separately</a></li>
</ul>
</div>

<div id="pca-with-known-minor-allele-and-snp-calling" class="section level2">
<h2>PCA with known minor allele and SNP calling</h2>
<p>SNP calling has identified 73,841 variable sites.</p>
<pre class="r"><code>#list.files(getwd(), &quot;.*covar&quot;)
covar = as.matrix(read.table(&quot;ParEry.covar&quot;, header=F))</code></pre>
<pre class="r"><code># getting filenames for labelling:
filenames = scan(&quot;/data3/claudius/Big_Data/ANGSD/slim.bamfile.list&quot;, what=&quot;character&quot;)
(Names = gsub(&quot;Data/&quot;, &quot;&quot;, gsub(&quot;.sorted.*&quot;, &quot;&quot;, filenames)))</code></pre>
<pre><code>##  [1] &quot;ery_30-10&quot; &quot;ery_30-11&quot; &quot;ery_30-12&quot; &quot;ery_30-13&quot; &quot;ery_30-14&quot;
##  [6] &quot;ery_30-15&quot; &quot;ery_30-16&quot; &quot;ery_30-17&quot; &quot;ery_30-18&quot; &quot;ery_30-1&quot; 
## [11] &quot;ery_30-2&quot;  &quot;ery_30-3&quot;  &quot;ery_30-4&quot;  &quot;ery_30-5&quot;  &quot;ery_30-6&quot; 
## [16] &quot;ery_30-7&quot;  &quot;ery_30-8&quot;  &quot;ery_30-9&quot;  &quot;par_34-10&quot; &quot;par_34-11&quot;
## [21] &quot;par_34-12&quot; &quot;par_34-13&quot; &quot;par_34-14&quot; &quot;par_34-15&quot; &quot;par_34-16&quot;
## [26] &quot;par_34-17&quot; &quot;par_34-18&quot; &quot;par_34-1&quot;  &quot;par_34-2&quot;  &quot;par_34-3&quot; 
## [31] &quot;par_34-4&quot;  &quot;par_34-5&quot;  &quot;par_34-6&quot;  &quot;par_34-7&quot;  &quot;par_34-8&quot; 
## [36] &quot;par_34-9&quot;</code></pre>
<pre class="r"><code># PCA using &#39;prcomp&#39; function
pca = prcomp(covar)
#summary(pca)
#names(pca)
#pca$sdev
plot(pca)</code></pre>
<p><img src="PCA_files/figure-html/unnamed-chunk-6-1.png" title="" alt="" width="672" /></p>
<pre class="r"><code>pc1_prop_var = (pca$sdev[1]^2)/sum((pca$sdev)^2)
pc2_prop_var = (pca$sdev[2]^2)/sum((pca$sdev)^2)</code></pre>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,2], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC2 (&quot;, signif(pc2_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;73,841 variable sites\n known minor allele + SNP calling&quot;
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,2])
pc2_sd = (pca$rot[,2] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,2][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/knownMinor_SNPcalling-1.png" title="" alt="" width="672" /></p>
</div>
<div id="pca-with-unknown-minor-allele-and-snp-calling" class="section level2">
<h2>PCA with unknown minor allele and SNP calling</h2>
<p>SNP calling has identified 68,590 variable sites.</p>
<pre class="r"><code>#list.files(getwd())
covar = as.matrix(read.table(&quot;withSNPcall_unknownMinor/EryPar.covar&quot;, header=F))</code></pre>
<pre class="r"><code># PCA using &#39;prcomp&#39; function
pca = prcomp(covar)
#summary(pca)
plot(pca)</code></pre>
<p><img src="PCA_files/figure-html/unnamed-chunk-8-1.png" title="" alt="" width="672" /></p>
<pre class="r"><code>pc1_prop_var = (pca$sdev[1]^2)/sum((pca$sdev)^2)
pc2_prop_var = (pca$sdev[2]^2)/sum((pca$sdev)^2)
pc3_prop_var = (pca$sdev[3]^2)/sum((pca$sdev)^2)</code></pre>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,2], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC2 (&quot;, signif(pc2_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;68,590 variable sites\n unknown minor allele + SNP calling&quot;
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,2])
pc2_sd = (pca$rot[,2] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[4]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,2][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/UNknownMinor_SNPcalling-1.png" title="" alt="" width="672" /></p>
<p>When taking the uncertainty in the minor allele into account, the 2nd principal component picks up different variation in the covariance matrix than when using the ML minor allele (above PCA). I wonder whether this difference is significant.</p>
<p>Let’s plot PC1 versus PC3 instead:</p>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,3], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC3 (&quot;, signif(pc3_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;68,590 variable sites\n unknown minor allele + SNP calling&quot;
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,3])
pc3_sd = (pca$rot[,3] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc3_sd &gt;= sort(pc3_sd, dec=TRUE)[4]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,3][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/UNknownMinor_SNPcalling-PC3-1.png" title="" alt="" width="672" /></p>
<p>So the PC3 picks up the variation that is picked up by PC2 in the other two PCA’s.</p>
</div>
<div id="pca-with-unknown-minor-allele-snp-calling-genotype-calling" class="section level2">
<h2>PCA with unknown minor allele + SNP calling + genotype calling</h2>
<pre class="r"><code>#list.files(getwd())
covar = as.matrix(read.table(&quot;withSNPcall_unknownMinor/EryPar.covar.GC&quot;, header=F))</code></pre>
<pre class="r"><code># PCA using &#39;prcomp&#39; function
pca = prcomp(covar)
#summary(pca)
plot(pca)</code></pre>
<p><img src="PCA_files/figure-html/unnamed-chunk-11-1.png" title="" alt="" width="672" /></p>
<pre class="r"><code>pc1_prop_var = (pca$sdev[1]^2)/sum((pca$sdev)^2)
pc2_prop_var = (pca$sdev[2]^2)/sum((pca$sdev)^2)
pc3_prop_var = (pca$sdev[3]^2)/sum((pca$sdev)^2)</code></pre>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,2], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC2 (&quot;, signif(pc2_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;68,590 variable sites\n unknown minor allele + SNP calling + genotype calling&quot;,
     cex.main=.9
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,3])
pc2_sd = (pca$rot[,2] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[4]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,2][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/UNknownMinor_SNPcalling_GenoCall-1.png" title="" alt="" width="672" /></p>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,3], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC3 (&quot;, signif(pc3_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;68,590 variable sites\n unknown minor allele + SNP calling + genotype calling&quot;,
     cex.main=.9
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,3])
pc3_sd = (pca$rot[,3] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc3_sd &gt;= sort(pc3_sd, dec=TRUE)[4]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,3][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/UNknownMinor_SNPcalling_GenoCall-PC3-1.png" title="" alt="" width="672" /></p>
</div>
<div id="pca-with-unknown-minor-allele-and-without-snp-calling" class="section level2">
<h2>PCA with unknown minor allele and without SNP calling</h2>
<p>This covariance matrix was created from 1,730,389 sites while weighting each site by its probability of being variable (eq. 20 in Fumagalli2013).</p>
<pre class="r"><code>#list.files(getwd())
covar = as.matrix(read.table(&quot;noSNPcall_unknownMinor/EryPar.covar&quot;, header=F))</code></pre>
<pre class="r"><code># PCA using &#39;prcomp&#39; function
pca = prcomp(covar)
#summary(pca)
plot(pca)</code></pre>
<p><img src="PCA_files/figure-html/unnamed-chunk-13-1.png" title="" alt="" width="672" /></p>
<pre class="r"><code>pc1_prop_var = (pca$sdev[1]^2)/sum((pca$sdev)^2)
pc2_prop_var = (pca$sdev[2]^2)/sum((pca$sdev)^2)</code></pre>
<pre class="r"><code>par(mfrow=c(1,1))
plot(pca$rot[,1], pca$rot[,2], xlim=c(-0.3, 0.3), 
     xlab=paste(&quot;PC1 (&quot;, signif(pc1_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     ylab=paste(&quot;PC2 (&quot;, signif(pc2_prop_var, digits=2)*100, &quot;%)&quot;, sep=&quot;&quot;), 
     col=c(rep(&quot;red&quot;, 18), rep(&quot;green&quot;, 18)), pch=16,
     main=&quot;1,730,389 sites\n unknown minor allele + weighting by p(var)&quot;,
     cex.main=.9
    )
legend(&quot;topleft&quot;, 
       legend=c(&quot;ERY&quot;, &quot;PAR&quot;),
       pch=16,
       col=c(&quot;red&quot;, &quot;green&quot;),
       bty=&quot;n&quot;
       )
# get indeces for outliers:
mu = mean(pca$rot[,1])
pc1_sd = (pca$rot[,1] - mu)^2
#Names[order(pc1_sd)]
#Names[which(pc1_sd &lt;= sort(pc1_sd)[2])] # two closest closest inds on PC1
index = which(pc1_sd &lt;= sort(pc1_sd)[2]) 
mu2 = mean(pca$rot[,2])
pc2_sd = (pca$rot[,2] - mu2)^2
#Names[order(pc2_sd, decreasing=TRUE)]
#Names[which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3])] # three most distant inds to mean of PC2
index = c(index, which(pc2_sd &gt;= sort(pc2_sd, dec=TRUE)[3]))
# label outliers:
text(pca$rot[,1][index], pca$rot[,2][index], labels=Names[index], cex=0.7, adj=c(0,0))</code></pre>
<p><img src="PCA_files/figure-html/UNknownMinor_NOsnpCalling-1.png" title="" alt="" width="672" /></p>
<p>The PCA plot is just flipped on the PC2 axis compared to the PCA with known minor allele and SNP calling. The signs of eigen vectors are arbitrary. So this flipping bears no meaning. Otherwise, this PCA looks identical to the first, with known minor allele and SNP calling, but different PC2 compared to the second PCA, with unknown minor allele and SNP calling.</p>
</div>
<div id="ery-and-par-separately" class="section level2">
<h2>ERY and PAR separately</h2>
<pre class="r"><code>par(mfrow=c(1,2))
#
pca_ery = prcomp(read.table(&quot;just_ERY/Ery.covar&quot;, header=F))
plot(pca_ery$rot[,1], pca_ery$rot[,2], 
     xlim=c(1.2*min(pca_ery$rot[,1]), 1.2*max(pca_ery$rot[,1])), 
     ylim=c(1.2*min(pca_ery$rot[,2]), 1.2*max(pca_ery$rot[,2])), 
     xlab=&quot;PC1&quot;, ylab=&quot;PC2&quot;, pch=16, col=&quot;red&quot;, main=&quot;ERY&quot;)
text(pca_ery$rot[,1], pca_ery$rot[,2], labels=Names[1:18], cex=.6, adj=c(1,1))
#
pca_par = prcomp(read.table(&quot;just_PAR/Par.covar&quot;, header=F))
plot(pca_par$rot[,1], pca_par$rot[,2], 
     xlim=c(1.2*min(pca_par$rot[,1]), 1.2*max(pca_par$rot[,1])), 
     ylim=c(1.2*min(pca_par$rot[,2]), 1.2*max(pca_par$rot[,2])),
     xlab=&quot;PC1&quot;, ylab=&quot;PC2&quot;, pch=16, col=&quot;green&quot;, main=&quot;PAR&quot;)
text(pca_par$rot[,1], pca_par$rot[,2], labels=Names[19:36], cex=.6, adj=c(1,1))</code></pre>
<p><img src="PCA_files/figure-html/per-pop-PCA-1.png" title="" alt="" width="672" /></p>
<p>The individuals par_34-5, par_34-14 and par_34-7 are also outliers in the within population PCA. So I am fairly certain I got the labeling right.</p>
</div>


</div>

<script>

// add bootstrap table styles to pandoc tables
$(document).ready(function () {
  $('tr.header').parent('thead').parent('table').addClass('table table-condensed');
});

</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdn.mathjax.org/mathjax/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
