---
title: "Big_Data"
author: "Claudius"
date: "01/11/2016"
output:
  github_document:
    toc: yes
#  html_document:
#    self_contained: false
#    theme: cosmo
#    toc: true
---

```{r setup options, cache=FALSE, include=FALSE}
library(knitr)
opts_chunk$set(dev=c('png', 'pdf'))
setwd("/data3/claudius/Big_Data/BOWTIE2/BAM")
```

I have created two coverage files with `bamtools coverage` from the BAM files of two individuals that shall
represent their respective population:

```{r}
infiles = Sys.glob("*cov.gz")
```

```{r, cache=TRUE}
ery = as.numeric( scan(pipe(paste("gzip -dc", infiles[1], "| cut -f 3"), open="r")) )
ery[1:10]
max(ery)
min(ery)
length(ery)
```

```{r}
par = as.numeric( scan(pipe(paste("gzip -dc", infiles[2], "| cut -f 3"), open="r")) )
par[1:10]
max(par)
min(par)
length(par)
```

I have coverage information from `r length(ery)` sites for the individual ery_30-15 and `r length(par)` for individual par_34-14.

```{r cov-dist-ery}
hist(ery,
     breaks=max(ery)/2,
     xlab="read coverage",
     ylab="frequency",
     freq=FALSE,
     xlim=c(0,100),
     main="Distribution of coverage for ind ery_30-15",
     col="blue"
     )

( Q = quantile(ery, probs=c(0.25, 0.5, 0.75, 0.95, 0.99, 0.999)) )

abline(v=Q["50%"], col="darkgreen", lty="solid", lwd=2)
abline(v=Q["95%"], col="red", lty="solid", lwd=2)

legend("topright",
       legend=c("median", "95th %"),
       title="Quantiles",
       col=c("darkgreen", "red"),
       lty="solid"
       )
```


```{r}
avg_cov = mean(ery)
avg_cov + 4*sqrt(avg_cov)
```

```{r cov-dist-par, echo=FALSE}
hist(par,
     breaks=max(par)/2,
     xlab="read coverage",
     ylab="frequency",
     freq=FALSE,
     xlim=c(0,100),
     main="Distribution of coverage for ind par_34-14",
     col="blue"
     )

( Q = quantile(par, probs=c(0.25, 0.5, 0.75, 0.95, 0.99, 0.999)) )

abline(v=Q["50%"], col="darkgreen", lty="solid", lwd=2)
abline(v=Q["95%"], col="red", lty="solid", lwd=2)
#
legend("topright",
       legend=c("median", "95th %"),
       title="Quantiles",
       col=c("darkgreen", "red"),
       lty="solid"
       )
```


## Depth distribution over SE RAD tags

I have created coverage count files from SE reads mapping to the 2nd position in the contigs (see assembly.sh).

```{r cov-dist-all, cache=TRUE, warning=FALSE}
infiles = list.files(getwd(), pattern=".*SE_depths")
depths = list()
for (i in 1:length(infiles)){
  cat( infiles[i], "\n" )
  depths[[i]] = scan(pipe(paste("cut -f 2", infiles[i]), open="r"))
}
x = depths[[1]]
plot(table(x)[1:50]/length(x), 
     ylim=c(0,0.25), 
     type="l",
     xlab="coverage",
     ylab="frequency among all contigs",
     main="coverage distributions of all 36 ind")
for(i in 2:length(depths)){
  x = depths[[i]]
  lines(1:50, table(x)[1:50]/length(x), type="l", col=i, lwd=2)
}
```

```{r}
Q99 = numeric(length(depths))
Q95 = numeric(length(depths))
for(i in 1:length(depths)){
  cat(infiles[i], "\t")
  Q99[i] = quantile(depths[[i]], probs=0.99)
  Q95[i] = quantile(depths[[i]], probs=0.95)
  cat(Q95[i], "\t", Q99[i], "\n")
}
summary(Q95)
summary(Q99)
# number and proportion of discarded contigs with maxDepth 25x:
sum(depths[[1]] > 25)
sum(depths[[1]] > 25)/length(depths[[1]])
# number and proportion of discarded contigs with maxDepth 101x:
sum(depths[[1]] > 101)
sum(depths[[1]] > 101)/length(depths[[1]])
```




## Mapping quality distribution

Note, after bowtie2 mapping, I discarded all mappings with quality 0.

```{r mapping-qual-dist, cache=TRUE}
infiles = list.files(getwd(), pattern="*sorted.bam$")
infiles
mq = scan(pipe(paste("samtools view", infiles[1], "| cut -f 5"), open="r"))
# hist(mq, breaks=max(mq), col="blue", xlim=c(0,45)) # produces rubbish, for unknown reason
c = tabulate(mq)
cbind(1:length(c), c, c/length(mq))
mp = barplot(c/length(mq),
        ylim=c(0,.3), 
        col=c("blue"),
        xlab="mapQ",
        ylab="proportion of reads",
        main=paste("mapping quality dist for", infiles[1])
        )
axis(side=1, 
     at=mp[c(1, 10, 20, 30, 40)], 
     labels=as.character(c(1, 10, 20, 30, 40))
     )
```

```{r}
sum(mq < 10)/length(mq)
```

About 40% of reads have a mapping quality below 10.

I would like to get the distribution of mapping qualities of all mapped reads before any filtering, for ERY and PAR separately. Then I would like to get the distribution of mapping qualities of all mapped reads after applying all filters (i. e. from those reads that overlap with keep.sites), for ERY and PAR separately. 

```{r, eval=FALSE}
# get get mapQ dist before filtering
# PAR
setwd("/data3/claudius/Big_Data/ANGSD/Data")
pp = pipe("for BAM in par*sorted.bam; do samtools view $BAM | cut -f 5; done", open="r")
mq.par = tabulate(scan(pp))
close(pp)

save(mq.par, file = "/data3/claudius/Big_Data/ANGSD/Data/mapQ_PAR.RData")
```

```{r, eval=FALSE}
# get get mapQ dist before filtering
# ERY
pp = pipe("for BAM in ery*sorted.bam; do samtools view $BAM | cut -f 5; done", open="r")
mq.ery = tabulate(scan(pp))
close(pp)

save(mq.ery, file = "/data3/claudius/Big_Data/ANGSD/Data/mapQ_ERY.RData")
```


```{r}
opts_chunk$set(root.dir="/data3/claudius/Big_Data/ANGSD/Data")
```

```{r}
load("/data3/claudius/Big_Data/ANGSD/Data/mapQ_PAR.RData")
cbind(1:length(mq.par), mq.par, mq.par/sum(mq.par))
mp = barplot(mq.par/sum(mq.par),
        ylim=c(0,.3), 
        col=c("blue"),
        xlab="mapping quality score",
        ylab="proportion of reads",
        main="mapping quality distribution for mapped reads of PAR"
        )
axis(side=1, 
     at=mp[c(1, 10, 20, 30, 40)], 
     labels=as.character(c(1, 10, 20, 30, 40))
     )
```





```{r}
load("/data3/claudius/Big_Data/ANGSD/Data/mapQ_ERY.RData")
mp = barplot(mq.ery/sum(mq.ery),
        ylim=c(0,.3), 
        col=c("blue"),
        xlab="mapping quality score",
        ylab="proportion of reads",
        main="mapping quality distribution for mapped reads of ERY"
        )
axis(side=1, 
     at=mp[c(1, 10, 20, 30, 40)], 
     labels=as.character(c(1, 10, 20, 30, 40))
     )
```

The file `ParEry.noSEgt2.nogtQ99Cov.noDUST.3.15.noTGCAGG.ANGSD_combinedNegFisFiltered.noGtQ99GlobalCov.sorted.sites` lists all sites in the Big Data reference that have passed my filters (see line 1308 onwards in `assembly.sh`). With my script `sites2bed.pl` I have turned this file into bed format. I have created a symbolic link to this file with the name `keep.bed`.

```{r, eval=FALSE}
setwd("/data3/claudius/Big_Data/ANGSD/Data")
# get get mapQ dist after filtering
# PAR
pp = pipe("for BAM in par*sorted.bam; do samtools view -L keep.bed $BAM | cut -f 5; done", open="r")
mq.after.par = tabulate(scan(pp))
close(pp)
save(mq.after.par, file = "/data3/claudius/Big_Data/ANGSD/Data/mapQ_PAR_afterFiltering.RData")

# get get mapQ dist after filtering
# ERY
pp = pipe("for BAM in ery*sorted.bam; do samtools view -L keep.bed $BAM | cut -f 5; done", open="r")
mq.after.ery = tabulate(scan(pp))
close(pp)
save(mq.after.ery, file = "/data3/claudius/Big_Data/ANGSD/Data/mapQ_ERY_afterFiltering.RData")
```


```{r, fig.width=8}
load("/data3/claudius/Big_Data/ANGSD/Data/mapQ_PAR_afterFiltering.RData")
# PAR
par.mq = as.matrix(rbind(mq.par/sum(mq.par), mq.after.par/sum(mq.after.par)))
rownames(par.mq) = c("before filtering", "after filtering")

mp = barplot(par.mq,
        ylim=c(0,.3), 
        xlab="mapping quality score",
        ylab="proportion of reads",
        main="mapping quality distribution for PAR",
        col=c("orange", "blue"), 
        beside=TRUE,
        legend=T,
        args.legend = list(x = "top")
        )
axis(side=1, 
     at=mp[c(1, 10, 20, 30, 40)*2]-mp[1]/2, 
     labels=as.character(c(1, 10, 20, 30, 40))
     )
```

```{r, fig.width=8}
load("/data3/claudius/Big_Data/ANGSD/Data/mapQ_ERY_afterFiltering.RData")
# ERY
ery.mq = as.matrix(rbind(mq.ery/sum(mq.ery), mq.after.ery/sum(mq.after.ery)))
rownames(ery.mq) = c("before filtering", "after filtering")

mp = barplot(ery.mq,
        ylim=c(0,.3), 
        xlab="mapping quality score",
        ylab="proportion of reads",
        main="mapping quality distribution for ERY",
        col=c("orange", "blue"), 
        beside=TRUE,
        legend=T,
        args.legend = list(x="top")
        )
axis(side=1, 
     at=mp[c(1, 10, 20, 30, 40)*2]-mp[1]/2, 
     labels=as.character(c(1, 10, 20, 30, 40))
     )
```

Filtering the reference sequence has selected for contigs with more unique sequences as shown by the great reduction in the proportion of reads with a mapping quality of 1 that map to the filtered reference as compared to the unfiltered reference sequence. Only reads with mapQ $\ge$ 5 have been used for downstream analyses.


## Contig length distribution

I would like to see the distribution of contig lengths before and after filtering. For commands on how I extracted the contig lengths, see `assembly.sh` line 2865.

```{r new setup options, cache=FALSE, include=FALSE}
opts_knit$set(root.dir = "/data3/claudius/Big_Data/ANGSD/Data")
setwd("/data3/claudius/Big_Data/ANGSD/Data")
dir("/data3/claudius/Big_Data/ANGSD/Data")
```

```{r}
# get contig lengths before filtering
setwd("/data3/claudius/Big_Data/ANGSD/Data")
before = read.delim("Big_Data_ref_contig_lengths", header=F)
names(before) = list("contig", "length")
head(before)
```

```{r}
# contig lengths after filtering
setwd("/data3/claudius/Big_Data/ANGSD/Data")
after = read.delim("keep.contig_lengths", header=F, sep=" ")
names(after) = list("contig", "length")
head(after)
```

```{r contig-length-dist, fig.height=12, fig.width=10}
par(mfrow=c(2,1))
#hist(before$length)
#hist(after$length)
plot(tabulate(before$length), type="l", ylab="contig count", xlab="contig length", xaxt="n")
axis(1, at=seq(0, 800, 100), label=seq(0, 800, 100))
plot(tabulate(after$length), type="l", ylab="contig count", xlab="contig length", xaxt="n")
axis(1, at=seq(0, 800, 100), label=seq(0, 800, 100))
```